// @nylas/web-elements/dist/collection/components/scheduler/nylas-scheduler/nylas-scheduling.js
import { NylasBaseProvider } from "../../../common/abstract-provider";
import { NylasSchedulerConnector } from "../../../connector/nylas-scheduler-connector/index";
import { CreateNylasSchedulerStore } from "../../../stores/scheduler-store";
import { NotificationType } from "@nylas/core";
import { Host, h } from "@stencil/core";
import { addDaysToCurrentDate, debug } from "../../../utils/utils";
import { SchedulerView } from "./scheduler-view";
import { ErrorCategory } from "../../../connector/nylas-scheduler-connector/errors/index";
import { MessageBanner } from "./message-banner";
const genericError = { title: 'Something went wrong...', type: NotificationType.Error, category: ErrorCategory.Component };
export class NylasScheduling {
    constructor() {
        this.mode = 'app';
        this.defaultSchedulerState = undefined;
        this.sessionId = undefined;
        this.schedulerApiUrl = 'https://api.us.nylas.com';
        this.bookingInfo = undefined;
        this.rescheduleBookingId = undefined;
        this.cancelBookingId = undefined;
        this.configurationId = undefined;
        this.nylasBranding = true;
        this.eventOverrides = {};
        this.isLoading = undefined;
        this.themeConfig = undefined;
        this.showNotification = true;
        this.automaticComponentRegistration = true;
        this.selectedDateLabel = new Date().toLocaleDateString(undefined, { dateStyle: 'full' }) || 'No date selected';
    }
    connectedCallback() {
        debug(`[nylas-scheduler] connectedCallback`);
    }
    async componentWillLoad() {
        debug(`[nylas-scheduler] Component will load`);
        const schedulerWillLoadEvent = this.schedulerWillLoad.emit(this.host);
        if (schedulerWillLoadEvent.defaultPrevented) {
            return;
        }
        if (!this.checkIfSessionIdOrConfigIdExists()) {
            return;
        }
        this.stores = {
            scheduler: CreateNylasSchedulerStore({
                bookingInfo: this.bookingInfo,
                nylasBranding: this.nylasBranding,
                ...this.defaultSchedulerState,
            }),
        };
        this.baseProvider = new NylasBaseProvider(this.host, this.stores, this.automaticComponentRegistration, this.eventOverrides);
        const nylasSchedulerStore = this.baseProvider?.getStore('scheduler');
        this.nylasSchedulerConnector = new NylasSchedulerConnector({
            schedulerAPIURL: this.schedulerApiUrl,
            sessionId: this.sessionId,
            configId: this.configurationId,
            nylasSchedulerStore,
        });
        this.init.emit(this.host);
        this.baseProvider?.componentWillLoad(this.nylasSchedulerConnector);
        if (this.themeConfig) {
            for (const [key, value] of Object.entries(this.themeConfig)) {
                this.host.style.setProperty(`${key}`, value);
            }
        }
    }
    async componentDidLoad() {
        this.baseProvider?.componentDidLoad();
        if (!this.checkIfSessionIdOrConfigIdExists()) {
            return;
        }
        const settings = await this.nylasSchedulerConnector?.scheduler.getUISettings();
        if (!settings || 'error' in settings) {
            this.nylasSchedulerError.emit({ ...this.getErrorObject(settings?.error) });
        }
        const nylasSchedulerStore = this.baseProvider?.getStore('scheduler');
        const availableDaysInFuture = nylasSchedulerStore?.get('configSettings')?.scheduler?.available_days_in_future;
        const selectedTimeslot = nylasSchedulerStore?.get('selectedTimeslot');
        if (selectedTimeslot && selectedTimeslot?.start_time) {
            if (!this.isTimeslotValid(selectedTimeslot)) {
                return;
            }
            if (nylasSchedulerStore?.get('showBookingForm')) {
                nylasSchedulerStore?.set('selectedDate', selectedTimeslot.start_time);
                return;
            }
        }
        const selectedDate = nylasSchedulerStore?.get('selectedDate');
        const availability = nylasSchedulerStore?.get('availability');
        let result;
        if (availability && availability.length > 0) {
            return;
        }
        if (!selectedDate) {
            const today = new Date();
            const { startTime, endTime, endTimeForAvailableDaysInFuture } = this.validateAvailableDaysInFuture(today, availableDaysInFuture || 30);
            if (today.getTime() < endTimeForAvailableDaysInFuture * 1000) {
                result = await this.nylasSchedulerConnector?.scheduler.getAvailability(startTime, endTime);
            }
        }
        else {
            const { startTime, endTime, endTimeForAvailableDaysInFuture } = this.validateAvailableDaysInFuture(selectedDate, availableDaysInFuture || 30);
            if (selectedDate.getTime() < endTimeForAvailableDaysInFuture * 1000) {
                result = await this.nylasSchedulerConnector?.scheduler.getAvailability(startTime, endTime);
            }
        }
        if (!result || 'error' in result) {
            this.nylasSchedulerError.emit({ ...this.getErrorObject(result?.error) });
        }
        if (!selectedDate) {
            let _selectedDate = new Date();
            const firstAvailableDate = nylasSchedulerStore?.get('availability').find((timeslot) => new Date(timeslot.start_time) > new Date());
            if (firstAvailableDate) {
                _selectedDate = firstAvailableDate.start_time;
            }
            nylasSchedulerStore?.set('selectedDate', _selectedDate);
        }
    }
    componentDisconnected() {
        this.baseProvider?.componentDisconnected();
    }
    bookingInfoChanged(newVal) {
        this.stores?.scheduler.set('bookingInfo', newVal);
    }
    async registerComponentHandler(event) {
        this.baseProvider?.registerComponent(event.detail);
    }
    async unregisterComponentHandler(event) {
        this.baseProvider?.unregisterComponent(event.detail);
    }
    async monthChangedHandler(event) {
        const date = event.detail;
        const availableDaysInFuture = this.nylasSchedulerConnector?.schedulerStore?.get('configSettings')?.scheduler?.available_days_in_future;
        const { startTime, endTime, endTimeForAvailableDaysInFuture } = this.validateAvailableDaysInFuture(date, availableDaysInFuture || 30);
        if (startTime < endTimeForAvailableDaysInFuture && date.getTime() < endTimeForAvailableDaysInFuture * 1000) {
            const result = await this.nylasSchedulerConnector?.scheduler.getAvailability(startTime, endTime);
            if (!result || 'error' in result) {
                this.nylasSchedulerError.emit({ ...this.getErrorObject(result?.error) });
            }
        }
    }
    async bookButtonClickedHandler(event) {
        if (this.rescheduleBookingId) {
            const result = await this.nylasSchedulerConnector?.scheduler.rescheduleBooking(this.rescheduleBookingId, event.detail);
            if (!result || 'error' in result) {
                this.nylasSchedulerError.emit({ ...this.getErrorObject(result?.error) });
            }
            return;
        }
        const result = await this.nylasSchedulerConnector?.scheduler.bookTimeslot(event.detail);
        if (!result || 'error' in result) {
            this.nylasSchedulerError.emit({ ...this.getErrorObject(result?.error) });
        }
    }
    dateSelectedHandler(event) {
        const date = event.detail;
        this.selectedDateLabel = date ? date.toLocaleDateString(undefined, { dateStyle: 'full' }) : 'No date selected';
    }
    cancelBookedEventErrorHandler(event) {
        this.nylasSchedulerError.emit({ ...this.getErrorObject(event.detail.error) });
    }
    cancelBookedEventValidationErrorHandler(event) {
        this.nylasSchedulerError.emit({ ...this.getErrorObject(event.detail.error) });
    }
    rescheduleBookedEventErrorHandler(event) {
        this.nylasSchedulerError.emit({ ...this.getErrorObject(event.detail.error) });
    }
    async getNylasSchedulerStore() {
        return this.baseProvider?.getStore('scheduler');
    }
    async getNylasSchedulerConnector() {
        return this.nylasSchedulerConnector;
    }
    async getRef() {
        return this.host;
    }
    getErrorObject(error) {
        return {
            title: error?.title || genericError.title,
            category: error?.category || genericError.category,
            type: NotificationType.Error,
            description: error?.message || '',
            id: new Date().getTime().toString(),
            ttl: 'none',
        };
    }
    checkIfSessionIdOrConfigIdExists() {
        if (!this.sessionId && !this.configurationId) {
            this.nylasSchedulerError.emit({
                title: 'No sessionId!',
                category: ErrorCategory.Component,
                type: NotificationType.Error,
                description: 'Session ID is required',
                id: new Date().getTime().toString(),
                ttl: 'none',
            });
            return false;
        }
        return true;
    }
    isTimeslotValid(timeslot) {
        const selectedDate = new Date(timeslot.start_time);
        if (selectedDate < new Date()) {
            this.nylasSchedulerError.emit({
                title: 'Invalid tiemslot!',
                category: ErrorCategory.Component,
                type: NotificationType.Error,
                description: 'The selected timeslot is invalid.',
                id: new Date().getTime().toString(),
                ttl: 'none',
            });
            return false;
        }
        return true;
    }
    validateAvailableDaysInFuture(startDate, availableDaysInFuture) {
        const today = new Date();
        const startTime = new Date(startDate.getFullYear(), startDate.getMonth(), 1).getTime() / 1000;
        const endTimeForCurrentMonth = new Date(startDate.getFullYear(), startDate.getMonth() + 1, 1).getTime() / 1000;
        const endTimeForAvailableDaysInFuture = Math.floor(addDaysToCurrentDate(today, availableDaysInFuture).getTime() / 1000);
        const endTime = Math.min(endTimeForAvailableDaysInFuture, endTimeForCurrentMonth);
        return { startTime, endTime, endTimeForAvailableDaysInFuture };
    }
    render() {
        const showDefaultScheduler = this.mode === 'app';
        const nylasSchedulerStore = this.baseProvider?.getStore('scheduler');
        const eventInfo = nylasSchedulerStore?.get('eventInfo');
        const showBookingForm = nylasSchedulerStore?.get('showBookingForm');
        const cancelledEventInfo = nylasSchedulerStore?.get('cancelledEventInfo');
        const cancelBookingId = nylasSchedulerStore?.get('cancelBookingId') || this.cancelBookingId;
        const selectedTimeslot = nylasSchedulerStore?.get('selectedTimeslot');
        const rescheduleBookingId = nylasSchedulerStore?.get('rescheduleBookingId') || this.rescheduleBookingId;
        const nylasBranding = nylasSchedulerStore?.get('nylasBranding') || this.nylasBranding;
        if (showDefaultScheduler && cancelBookingId && !cancelledEventInfo) {
            return (h(Host, null, h(SchedulerView, { nylasBranding: nylasBranding, isLoading: this.isLoading, showNotification: this.showNotification }, h("div", { class: "cancel-flow-page", part: "cancel-flow-page" }, h("nylas-cancel-booking-form", { cancelBookingId: cancelBookingId, exportparts: "ncbf, ncbf__input-textfield, ncbf__button-cta, ncbf__button-outline, ncbf__input-label" })))));
        }
        if (showDefaultScheduler && cancelledEventInfo) {
            return (h(Host, null, h(SchedulerView, { nylasBranding: nylasBranding, isLoading: this.isLoading, showNotification: this.showNotification }, h("div", { class: "cancelled-event-page", part: "cancelled-event-page" }, h("nylas-cancelled-event-card", { cancelledEventInfo: cancelledEventInfo, exportparts: "ncec, ncec__icon, ncec__title, ncec__description, ncec__button-outline, ncec__card" })))));
        }
        if (showDefaultScheduler && eventInfo) {
            return (h(Host, null, h(SchedulerView, { nylasBranding: nylasBranding, isLoading: this.isLoading, showNotification: this.showNotification }, h("div", { class: "booked-event-page", part: "booked-event-page" }, h("nylas-booked-event-card", { selectedTimeslot: selectedTimeslot, eventInfo: eventInfo, exportparts: "nbec, nbec__title, nbec__card, nbec__description, nbec__button-outline, nbec__cancel-cta, nbec__reschedule-cta" })))));
        }
        if (showDefaultScheduler && (!eventInfo || !!rescheduleBookingId) && !showBookingForm) {
            const timeslotPickerCTALabel = this.host.querySelector('[slot="timeslot-picker-cta-label"]');
            return (h(Host, null, h(SchedulerView, { nylasBranding: nylasBranding, isLoading: this.isLoading, showNotification: this.showNotification }, rescheduleBookingId && h(MessageBanner, null, "Choose another time slot to reschedule your meeting"), h("div", { class: "select-date-page", part: "select-date-page" }, h("div", { class: "left-panel" }, h("nylas-date-picker", { exportparts: "ndp, ndp__month-header, ndp__month-button, ndp__day, ndp__date, ndp__date--selected, ndp__date--current-day, ndp__date--current-month" }), h("div", { class: "border" }), h("nylas-locale-switch", null)), h("div", { class: "right-panel" }, h("h2", null, h("calendar-icon", null), h("span", { id: "selectedDate" }, this.selectedDateLabel)), h("nylas-timeslot-picker", { exportparts: "ntp, ntp__timeslot, ntp__timeslot--selected, ntp__button-primary" }, timeslotPickerCTALabel && (h("span", { slot: "timeslot-picker-cta-label" }, h("slot", { name: "timeslot-picker-cta-label" })))))))));
        }
        if (showDefaultScheduler && !eventInfo && showBookingForm) {
            const customTemplateSlot = this.host.querySelector('[slot="custom-booking-form"]');
            return (h(Host, null, h(SchedulerView, { nylasBranding: nylasBranding, isLoading: this.isLoading, showNotification: this.showNotification }, rescheduleBookingId && h(MessageBanner, null, "Choose another time slot to reschedule your meeting"), h("div", { class: "additional-data-page", part: "additional-data-page" }, h("div", { class: "left-panel" }, h("div", { class: "wrapper" }, h("nylas-selected-event-card", { exportparts: "nsec, nsec__card, nsec__icon, nsec__date, nsec__time" }))), h("div", { class: "right-panel" }, h("nylas-booking-form", { exportparts: "nbf, nbf__input-label, nbf__input-textfield, nbf__button-ghost, nbf__button-primary, nbf__button-outline" }, customTemplateSlot && (h("div", { slot: "custom-booking-form" }, h("slot", { name: "custom-booking-form" })))))))));
        }
        return (h(Host, null, h("slot", null)));
    }
    static get is() { return "nylas-scheduling"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["nylas-scheduling.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["nylas-scheduling.css"]
        };
    }
    static get properties() {
        return {
            "mode": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'app' | 'composable'",
                    "resolved": "\"app\" | \"composable\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This is used to set the mode for the Nylas Scheduler.\nThe mode can be either `app` or `composable`. The default mode is `app`.\n- `app`: This mode is used to show the default Nylas Scheduler UI.\n- `composable`: This mode is used to show the composable Nylas Scheduler UI\n   by passing the individual scheduler components as children."
                },
                "attribute": "mode",
                "reflect": false,
                "defaultValue": "'app'"
            },
            "defaultSchedulerState": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Partial<NylasSchedulerStoreState>",
                    "resolved": "undefined | { selectedDate?: Date | null | undefined; selectedLanguage?: string | undefined; selectedTimezone?: string | undefined; selectedTimeslot?: Timeslot | null | undefined; showBookingForm?: boolean | undefined; selectableDates?: Date[] | null | undefined; availability?: AvailabilityTimeslot[] | undefined; state?: DataState | undefined; eventInfo?: NylasEvent | null | undefined; cancelledEventInfo?: Partial<NylasEvent> | null | undefined; bookingInfo?: NylasSchedulerBookingData | undefined; rescheduleBookingId?: string | undefined; cancelBookingId?: string | undefined; isLoading?: boolean | undefined; nylasBranding?: boolean | undefined; configSettings?: ConfigSettings | undefined; }",
                    "references": {
                        "Partial": {
                            "location": "global",
                            "id": "global::Partial"
                        },
                        "NylasSchedulerStoreState": {
                            "location": "import",
                            "path": "@/stores/scheduler-store",
                            "id": "src/stores/scheduler-store.ts::NylasSchedulerStoreState"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The default scheduler store state.\nUsed to set the initial state of the scheduler store."
                }
            },
            "sessionId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The session ID for the Nylas Scheduler.\nThis is used to authenticate the user."
                },
                "attribute": "session-id",
                "reflect": false
            },
            "schedulerApiUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The URL for the Nylas Scheduler API. (staging or production URL)"
                },
                "attribute": "scheduler-api-url",
                "reflect": false,
                "defaultValue": "'https://api.us.nylas.com'"
            },
            "bookingInfo": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "NylasSchedulerBookingData",
                    "resolved": "undefined | { primaryParticipant: NylasSchedulerBookingParticipant; startTime?: Date | undefined; endTime?: Date | undefined; timezone?: string | undefined; guests?: NylasSchedulerBookingParticipant[] | undefined; additionalFields?: { [key: string]: any; } | undefined; }",
                    "references": {
                        "NylasSchedulerBookingData": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::NylasSchedulerBookingData"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "This enables passing the bookingInfo object to the Scheduler UI for direct booking event creation.\nWhen used with timeslotConfirmedHandler, it bypasses the additional data page and immediately\ninvokes the handler after booking."
                }
            },
            "rescheduleBookingId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Booking Id required for rescheduling flow."
                },
                "attribute": "reschedule-booking-id",
                "reflect": false
            },
            "cancelBookingId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Booking Id required for cancelling flow."
                },
                "attribute": "cancel-booking-id",
                "reflect": false
            },
            "configurationId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The config ID for the Nylas Scheduler. This should be passed in when using a public config,\nin which case the sessionId is not required."
                },
                "attribute": "configuration-id",
                "reflect": false
            },
            "nylasBranding": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "This prop lets you hide the Nylas branding.\nDefault is true."
                },
                "attribute": "nylas-branding",
                "reflect": false,
                "defaultValue": "true"
            },
            "eventOverrides": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "SchedulerEventOverride",
                    "resolved": "Partial<{ timeslotSelected: (event: CustomEvent<Timeslot>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; timeslotConfirmed: (event: CustomEvent<Timeslot>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; dateSelected: (event: CustomEvent<Date>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; monthChanged: (event: CustomEvent<Date>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; timezoneChanged: (event: CustomEvent<string>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; languageChanged: (event: CustomEvent<string>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; bookingInfo: (event: CustomEvent<unknown>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; nameChanged: (event: CustomEvent<string>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; emailChanged: (event: CustomEvent<string>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; backButtonClicked: (event: CustomEvent<boolean>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; detailsConfirmed: (event: CustomEvent<NylasSchedulerBookingData>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; bookingFormError: (event: CustomEvent<Partial<Notification>>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; configSettings: (event: CustomEvent<unknown>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; cancelBookingButtonClicked: (event: CustomEvent<{ bookingId: string; }>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; rescheduleButtonClicked: (event: CustomEvent<{ bookingId: string; errorHandler?: ((error: NylasSchedulerErrorResponse) => void) | undefined; }>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; rescheduleBookedEventError: (event: CustomEvent<NylasSchedulerErrorResponse>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; cancelBookedEventValidationError: (event: CustomEvent<{ error: { title: string; message: string; }; }>, connector?: NylasSchedulerConnectorInterface | undefined) => Promise<void>; }> & EventOverride<NylasSchedulerConnectorInterface>",
                    "references": {
                        "SchedulerEventOverride": {
                            "location": "import",
                            "path": "@/common/component-types",
                            "id": "src/common/component-types.ts::SchedulerEventOverride"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This provides an easy way to override the default function of the event emitter.\nAn example of this is the `timeslotConfirmed` event. By default, this event will set the scheduler store state for `showBookingForm` to `true` which will\nshow the booking form. However, if you want to override this behavior, you can pass in the prop `eventOverride` like:\n```html\n<nylas-scheduling eventOverride={{\"timeslotConfirmed\": (event, nylasConnector) => { console.log(\"Timeslot confirmed event fired!\"); } }} />\n```"
                },
                "defaultValue": "{}"
            },
            "isLoading": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The loading state. This is used to set the loading state for the Nylas Scheduler when fetching data."
                },
                "attribute": "is-loading",
                "reflect": false
            },
            "themeConfig": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ThemeConfig",
                    "resolved": "undefined | { '--nylas-primary'?: string | undefined; '--nylas-info'?: string | undefined; '--nylas-success'?: string | undefined; '--nylas-warning'?: string | undefined; '--nylas-error'?: string | undefined; '--nylas-error-pressed'?: string | undefined; '--nylas-base-0'?: string | undefined; '--nylas-base-25'?: string | undefined; '--nylas-base-50'?: string | undefined; '--nylas-base-100'?: string | undefined; '--nylas-base-200'?: string | undefined; '--nylas-base-300'?: string | undefined; '--nylas-base-400'?: string | undefined; '--nylas-base-500'?: string | undefined; '--nylas-base-600'?: string | undefined; '--nylas-base-700'?: string | undefined; '--nylas-base-800'?: string | undefined; '--nylas-base-900'?: string | undefined; '--nylas-base-950'?: string | undefined; '--nylas-font-family'?: string | undefined; '--nylas-font-size'?: string | undefined; '--nylas-border-radius'?: string | undefined; '--nylas-border-radius-2x'?: string | undefined; '--nylas-border-radius-3x'?: string | undefined; }",
                    "references": {
                        "ThemeConfig": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::ThemeConfig"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Theme config, used to automatically generate a theme with color palette and\nCSS variables to customize the look and feel of the Nylas Scheduler."
                }
            },
            "showNotification": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "show-notification",
                "reflect": false,
                "defaultValue": "true"
            }
        };
    }
    static get states() {
        return {
            "automaticComponentRegistration": {},
            "selectedDateLabel": {}
        };
    }
    static get events() {
        return [{
                "method": "init",
                "name": "init",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the provider is initialized.\nIt can be used to set the initial state of the provider,\nor to prevent the provider from firing some default behavior."
                },
                "complexType": {
                    "original": "HTMLNylasSchedulingElement",
                    "resolved": "HTMLNylasSchedulingElement",
                    "references": {
                        "HTMLNylasSchedulingElement": {
                            "location": "global",
                            "id": "global::HTMLNylasSchedulingElement"
                        }
                    }
                }
            }, {
                "method": "schedulerWillLoad",
                "name": "schedulerWillLoad",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the scheduler component enters componentWillLoad lifecycle."
                },
                "complexType": {
                    "original": "HTMLNylasSchedulingElement",
                    "resolved": "HTMLNylasSchedulingElement",
                    "references": {
                        "HTMLNylasSchedulingElement": {
                            "location": "global",
                            "id": "global::HTMLNylasSchedulingElement"
                        }
                    }
                }
            }, {
                "method": "schedulerDidLoad",
                "name": "schedulerDidLoad",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the scheduler component enters componentDidLoad lifecycle."
                },
                "complexType": {
                    "original": "HTMLNylasSchedulingElement",
                    "resolved": "HTMLNylasSchedulingElement",
                    "references": {
                        "HTMLNylasSchedulingElement": {
                            "location": "global",
                            "id": "global::HTMLNylasSchedulingElement"
                        }
                    }
                }
            }, {
                "method": "nylasSchedulerError",
                "name": "nylasSchedulerError",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "Notification",
                    "resolved": "{ id: string; type: NotificationType; title: string; code?: number | undefined; category: string; description?: string | undefined; ttl?: number | \"none\" | undefined; }",
                    "references": {
                        "Notification": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::Notification"
                        }
                    }
                }
            }];
    }
    static get methods() {
        return {
            "getNylasSchedulerStore": {
                "complexType": {
                    "signature": "() => Promise<NylasSchedulerStoreType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasSchedulerStoreState": {
                            "location": "import",
                            "path": "@/stores/scheduler-store",
                            "id": "src/stores/scheduler-store.ts::NylasSchedulerStoreState"
                        },
                        "NylasSchedulerStoreType": {
                            "location": "import",
                            "path": "@/stores/scheduler-store",
                            "id": "src/stores/scheduler-store.ts::NylasSchedulerStoreType"
                        }
                    },
                    "return": "Promise<ObservableMap<NylasSchedulerStoreState> | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasScheduler instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasScheduler instance"
                        }]
                }
            },
            "getNylasSchedulerConnector": {
                "complexType": {
                    "signature": "() => Promise<NylasSchedulerConnector | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasSchedulerConnector": {
                            "location": "import",
                            "path": "@/connector/nylas-scheduler-connector",
                            "id": "src/connector/nylas-scheduler-connector/index.ts::NylasSchedulerConnector"
                        }
                    },
                    "return": "Promise<NylasSchedulerConnector | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasConnector instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasConnector instance"
                        }]
                }
            },
            "getRef": {
                "complexType": {
                    "signature": "() => Promise<HTMLNylasSchedulingElement>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLNylasSchedulingElement": {
                            "location": "global",
                            "id": "global::HTMLNylasSchedulingElement"
                        }
                    },
                    "return": "Promise<HTMLNylasSchedulingElement>"
                },
                "docs": {
                    "text": "",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "host"; }
    static get watchers() {
        return [{
                "propName": "bookingInfo",
                "methodName": "bookingInfoChanged"
            }];
    }
    static get listeners() {
        return [{
                "name": "registerComponent",
                "method": "registerComponentHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "unregisterComponent",
                "method": "unregisterComponentHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "monthChanged",
                "method": "monthChangedHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "detailsConfirmed",
                "method": "bookButtonClickedHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "dateSelected",
                "method": "dateSelectedHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "cancelBookedEventError",
                "method": "cancelBookedEventErrorHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "cancelBookedEventValidationError",
                "method": "cancelBookedEventValidationErrorHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "rescheduleBookedEventError",
                "method": "rescheduleBookedEventErrorHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
