// @nylas/web-elements/dist/collection/components/design-system/tooltip-component/toolitp-component.js
import { h } from "@stencil/core";
export class TooltipComponent {
    constructor() {
        this.position = 'bottom';
    }
    showTooltip(event) {
        event.target.nextElementSibling.style.visibility = 'visible';
        event.target.nextElementSibling.style.opacity = '1';
    }
    hideTooltip(event) {
        event.target.nextElementSibling.style.visibility = 'hidden';
        event.target.nextElementSibling.style.opacity = '0';
    }
    render() {
        return (h("div", { key: 'e5f137458d3664942d02b31f221b3b5ae2bd9658', class: "tooltip-container" }, h("span", { key: '4838a8e46e43d2f490c81381c3759142868ce45c', class: "info-icon", tabindex: "0", onFocus: this.showTooltip.bind(this), onBlur: this.hideTooltip.bind(this) }, h("slot", { key: 'a4864cffceabaf3bf353647733add76685ac1581', name: "tooltip-icon" }, "i")), h("span", { key: '15cbc68f4e0e07bddff0e7adbb76fd6ffce4b9a4', class: `tooltip-content ${this.position}` }, h("slot", { key: '836a48731f8a24bc4ecbbb097a514ca263dfc887', name: "tooltip-content" }, "This is a help text."))));
    }
    static get is() { return "tooltip-component"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tooltip-component.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tooltip-component.css"]
        };
    }
    static get properties() {
        return {
            "position": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'top' | 'bottom' | 'left' | 'right'",
                    "resolved": "\"bottom\" | \"left\" | \"right\" | \"top\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The position of the tooltip."
                },
                "attribute": "position",
                "reflect": false,
                "defaultValue": "'bottom'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
