// @nylas/web-elements/dist/collection/components/design-system/time-period-selector/time-period-selector.js
import { debug } from "../../../utils/utils";
import { h } from "@stencil/core";
export class TimePeriodSelector {
    constructor() {
        this.timePeriods = ['hour', 'day', 'week', 'month'];
        this.defaultSelectedPeriod = 'hour';
        this.defaultSelectedNumber = 1;
        this.selectedPeriod = this.defaultSelectedPeriod;
        this.selectedNumber = this.defaultSelectedNumber;
        this.numberOptions = this.calculateOptions('hour').map(i => {
            return {
                label: i.toString(),
                value: i.toString(),
            };
        });
        this.timePeriodOptions = this.timePeriods.map(period => {
            return {
                label: period,
                value: period,
            };
        });
    }
    defaultSelectedPeriodChangedHandler(newValue) {
        this.selectedPeriod = newValue;
        this.updateNumberOptionsAndSelectedNumber(newValue);
    }
    defaultSelectedNumberChangedHandler(newValue) {
        this.selectedNumber = newValue;
    }
    timePeriodsChangedHandler(newValue) {
        this.timePeriodOptions = newValue.map(period => {
            return {
                label: period,
                value: period,
            };
        });
    }
    calculateOptions(period) {
        switch (period) {
            case 'hour':
                return Array.from({ length: 23 }, (_, i) => i + 1);
            case 'minute':
                return Array.from({ length: 60 }, (_, i) => i);
            case 'day':
                return Array.from({ length: 30 }, (_, i) => i + 1);
            case 'week':
                return Array.from({ length: 4 }, (_, i) => i + 1);
            case 'month':
                return Array.from({ length: 12 }, (_, i) => i + 1);
            default:
                return [];
        }
    }
    selectedOptionChangedHandler(event) {
        debug('time-period-selector', 'selectedOptionChangedHandler', event.detail);
        const { value, name } = event.detail;
        if (name === 'time-period') {
            this.selectedPeriod = value;
            this.updateNumberOptionsAndSelectedNumber(value);
        }
        else if (name === 'time-number') {
            this.selectedNumber = parseInt(value);
        }
        const selected = {
            number: this.selectedNumber,
            period: this.selectedPeriod,
        };
        this.timePeriodChanged.emit(selected);
    }
    updateNumberOptionsAndSelectedNumber(period) {
        const numberOptions = this.calculateOptions(period);
        this.numberOptions = numberOptions.map(i => {
            return {
                label: i.toString(),
                value: i.toString(),
            };
        });
        this.selectedNumber = parseInt(this.numberOptions[0].value);
    }
    render() {
        return (h("div", { key: 'eaec00db41b0c2410fc4a9ccc341fc4ec4012e10', class: "time-period-selector" }, h("select-dropdown", { key: '0d53efe90429a05d4f52a8f1b92bf46b14b1e6ad', id: "time-number", name: 'time-number', exportparts: "sd__dropdown: tps__number-dropdown, sd__dropdown-button: tps__number-dropdown-button, sd__dropdown-content: tps__number-dropdown-content", options: this.numberOptions, defaultSelectedOption: this.numberOptions.find(i => i.value == this.selectedNumber.toString()) ?? this.numberOptions[0], withSearch: false }), h("select-dropdown", { key: 'bce764e3e959eb4fad73d1e410173d584f243ee5', id: "time-period", name: 'time-period', options: this.timePeriodOptions, exportparts: "sd__dropdown: tps__period-dropdown, sd__dropdown-button: tps__period-dropdown-button, sd__dropdown-content: tps__period-dropdown-content", pluralizedLabel: this.selectedNumber > 1 ? 's' : '', defaultSelectedOption: this.timePeriodOptions.find(i => i.value == this.selectedPeriod) ?? this.timePeriodOptions[0], withSearch: false })));
    }
    static get is() { return "time-period-selector"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() {
        return {
            "$": ["time-period-selector.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["time-period-selector.css"]
        };
    }
    static get properties() {
        return {
            "timePeriods": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "defaultValue": "['hour', 'day', 'week', 'month']"
            },
            "defaultSelectedPeriod": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The default selected time period."
                },
                "attribute": "default-selected-period",
                "reflect": false,
                "defaultValue": "'hour'"
            },
            "defaultSelectedNumber": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The default selected number."
                },
                "attribute": "default-selected-number",
                "reflect": false,
                "defaultValue": "1"
            }
        };
    }
    static get states() {
        return {
            "selectedPeriod": {},
            "selectedNumber": {},
            "numberOptions": {},
            "timePeriodOptions": {}
        };
    }
    static get events() {
        return [{
                "method": "timePeriodChanged",
                "name": "timePeriodChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "{ number: number; period: string }",
                    "resolved": "{ number: number; period: string; }",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "host"; }
    static get watchers() {
        return [{
                "propName": "defaultSelectedPeriod",
                "methodName": "defaultSelectedPeriodChangedHandler"
            }, {
                "propName": "defaultSelectedNumber",
                "methodName": "defaultSelectedNumberChangedHandler"
            }, {
                "propName": "timePeriods",
                "methodName": "timePeriodsChangedHandler"
            }];
    }
    static get listeners() {
        return [{
                "name": "selectedOptionChanged",
                "method": "selectedOptionChangedHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
