// @nylas/web-elements/dist/cjs/nylas-login.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c14ea8f5.js');
const registerComponent = require('./register-component-4f8ae42c.js');
require('./utils-85bd0909.js');
require('./_commonjsHelpers-ef8cd0cd.js');

const nylasLoginCss = ":host{font-family:'Arial', sans-serif;background:var(--nylas-color-primary-50);color:var(--nylas-color-primary-800);margin:0;display:contents}input{box-sizing:border-box}.login-container{display:flex;justify-content:center;align-items:center;height:100%}.login-box{width:300px;padding:20px;box-shadow:0 4px 8px rgba(0, 0, 0, 0.1);border-radius:var(--nylas-border-radius-2x);background:var(--nylas-color-primary-50);text-align:center}.login-box h1{margin-top:0;margin-bottom:0.5em}.login-box p{margin-bottom:1em}.login-methods button{display:block;width:100%;padding:10px 10px 10px 50px;margin-bottom:10px;border:none;border-radius:var(--nylas-border-radius);border:1px solid var(--nylas-color-primary-300);background:var(--nylas-color-primary-100);color:var(--nylas-color-primary-600);cursor:pointer;transition:background 0.3s ease;position:relative;text-align:left}.login-methods button svg{position:absolute;left:16px;top:50%;transform:translateY(-50%)}.login-methods button:hover{background:var(--nylas-color-primary-200);}.login-form label{display:block;margin-bottom:5px;text-align:left}.login-form input{width:100%;padding:10px;margin-bottom:10px;border:1px solid var(--nylas-color-primary-200);border-radius:var(--nylas-border-radius);background:var(--nylas-color-primary-100);color:var(--nylas-color-primary-800);}.login-form button{width:100%;padding:10px;border:none;border-radius:var(--nylas-border-radius);background:var(--nylas-color-primary-500);color:var(--nylas-color-primary-50);cursor:pointer;transition:background 0.3s ease}.login-form button:hover{background:var(--nylas-color-primary-600);}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasLogin = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.buttonClick = index.createEvent(this, "buttonClick", 7);
        this._handleLoginRedirect = async () => {
            const url = await this.nylasConnector?.auth.getHostedAuthRedirectURL(this.scopes);
            const event = this.buttonClick.emit({ url });
            if (event.defaultPrevented) {
                return;
            }
            if (url) {
                window.location.href = url;
            }
        };
        this._onGoogleButtonClick = async (event) => {
            event.preventDefault();
            const url = await this.nylasConnector?.auth.getHostedAuthRedirectURL(this.scopes, undefined, 'google');
            if (url) {
                window.location.href = url;
            }
        };
        this._onMicrosoftButtonClick = async (event) => {
            event.preventDefault();
            const url = await this.nylasConnector?.auth.getHostedAuthRedirectURL(this.scopes, undefined, 'microsoft');
            if (url) {
                window.location.href = url;
            }
        };
        this._onFormSubmit = async (event) => {
            event.preventDefault();
            const url = await this.nylasConnector?.auth.getHostedAuthRedirectURL(this.scopes, this.email);
            if (url) {
                window.location.href = url;
            }
        };
        this.isAuthenticated = false;
        this.nylasConnector = undefined;
        this.scopes = undefined;
        this.loginMethod = 'redirect';
        this.companyName = 'Nylas';
        this.logoUrl = 'https://www.nylas.com/wp-content/uploads/main-logo.svg';
        this.logoWidth = 'min-content';
        this.providers = 'google,microsoft';
        this.buttonText = 'Login';
        this.email = '';
    }
    connectedCallback() { }
    disconnectedCallback() { }
    componentWillLoad() { }
    async componentDidLoad() {
        const exchangedToken = await this.nylasConnector?.auth.manageHostedAuthCodeExchange(window.location.href);
        if (exchangedToken) {
            window.history.replaceState({}, document.title, window.location.pathname);
        }
    }
    render() {
        const splitProviders = this.providers?.split(',') || [];
        return (index.h(index.Host, { key: '124b2f41ebcd01e13b0c031a1240120835610fbb' }, this.loginMethod === 'redirect' && (index.h("div", { class: "redirect-container" }, !this.isAuthenticated && (index.h("sp-button", { part: "button", onClick: this._handleLoginRedirect }, this.buttonText)))), this.loginMethod === 'emebded' && (index.h("main", { class: "login-container" }, index.h("section", { class: "login-box" }, index.h("header", null, index.h("img", { src: this.logoUrl, alt: `${this.companyName} logo`, width: '100%', style: {
                minWidth: this.logoWidth,
            } }), index.h("h1", null, "Welcome")), index.h("p", null, "Log in to your account to continue to ", this.companyName), index.h("div", { class: "login-methods" }, splitProviders.includes('google') && (index.h("button", { class: "login-google", onClick: this._onGoogleButtonClick }, index.h("svg", { xmlns: "http://www.w3.org/2000/svg", "aria-hidden": "true", role: "img", width: "18", height: "18", preserveAspectRatio: "xMidYMid meet", viewBox: "0 0 256 262" }, index.h("path", { fill: "#4285F4", d: "M255.6 131.6c0-8.2-.8-16.2-2.2-23.9H130v45.3h70.3c-3 16.1-12.4 29.7-26.3 38.9v32.4h42.6c25-23 39.4-56.7 39.4-97.7z" }), index.h("path", { fill: "#34A853", d: "M130 262c35.3 0 64.9-11.7 86.5-31.7l-42.6-32.4c-11.7 7.9-26.7 12.6-43.9 12.6-33.7 0-62.3-22.7-72.5-53.4H14.5v33.3A130.3 130.3 0 00130 262z" }), index.h("path", { fill: "#FBBC05", d: "M57.5 158.1a72.7 72.7 0 010-54.2V70.6H14.5a130.4 130.4 0 000 120.8l43-33.3z" }), index.h("path", { fill: "#EA4335", d: "M130 51.6c19 0 36 6.5 49.4 19.2l37-36C195 13.1 165.2 0 130 0A130.3 130.3 0 0014.5 70.6l43 33.3c10.2-30.7 38.8-53.3 72.5-53.3z" })), "Continue with Google")), splitProviders.includes('microsoft') && (index.h("button", { class: "login-microsoft", onClick: this._onMicrosoftButtonClick }, index.h("svg", { xmlns: "http://www.w3.org/2000/svg", "aria-hidden": "true", role: "img", width: "18", height: "18", preserveAspectRatio: "xMidYMid meet", viewBox: "0 0 448 512" }, index.h("path", { fill: "#F3F3F3", d: "M0 32v192h192V32H0zm256 0v192h192V32H256zm192 256v192h-192V288h192zM0 224v192h192V224H0z" })), "Continue with Microsoft"))), index.h("form", { class: "login-form", method: "POST", onSubmit: this._onFormSubmit }, index.h("label", { htmlFor: "email" }, "Email"), index.h("input", { type: "email", id: "email", name: "email", required: true, value: this.email }), index.h("button", { type: "submit" }, "Next")))))));
    }
};
__decorate([
    registerComponent.RegisterComponent({
        name: 'nylas-login',
        connectorToProp: 'nylasConnector',
        stateToProps: new Map([['auth.isAuthenticated', 'isAuthenticated']]),
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasLogin.prototype, "render", null);
NylasLogin.style = nylasLoginCss;

exports.nylas_login = NylasLogin;
