// @nylas/web-elements/dist/cjs/nylas-notification.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c14ea8f5.js');
const utils = require('./utils-85bd0909.js');
require('./_commonjsHelpers-ef8cd0cd.js');

const nylasNotificationCss = ":host{display:block;margin-bottom:-12px;width:100%;z-index:1;--nylas-primary:#2563eb;--nylas-error:#cc4841;--nylas-error-pressed:#992222;--nylas-wraning:#f06c00;--nylas-success:#16a392;--nylas-info:#2b8fc2;--nylas-base-0:#ffffff;--nylas-base-25:#fcfcfd;--nylas-base-50:#f8f9fc;--nylas-base-100:#eaecf5;--nylas-base-200:#d5d9eb;--nylas-base-300:#b3b8d8;--nylas-base-400:#717bbc;--nylas-base-500:#4e5ba6;--nylas-base-600:#3e4784;--nylas-base-700:#263f72;--nylas-base-800:#293056;--nylas-base-900:#101323;--nylas-base-950:#0e101b;--nylas-border-radius:0.25rem;--nylas-border-radius-2x:0.5rem;--nylas-border-radius-3x:0.75rem;--nylas-font-family:\"Inter\", sans-serif}.notif{display:flex;padding:17px;align-items:flex-start;justify-content:space-between;gap:16px;border-radius:var(--nylas-border-radius-2x);border:1.5px solid var(--nylas-error);background:var(--nylas-base-0)}@media screen and (max-width: 768px){.notif{width:auto}}.notif .content{display:flex;flex-direction:row;gap:8px}.notif .content .message{font-size:14px;font-weight:400;color:var(--nylas-base-800)}.notif .content .message .title{font-size:16px;font-weight:600;color:var(--nylas-base-950);display:block}.notif .content .message .description{font-size:14px;font-weight:400;color:var(--nylas-base-800);display:block}.notif .content .icon.error{color:var(--nylas-error)}.notif.error{border-color:var(--nylas-error)}.notif.warning{border-color:var(--nylas-warning)}.notif.info{border-color:var(--nylas-info)}.notif.success{border-color:var(--nylas-success)}.notif .dismiss-btn{background:none;border:none;color:var(--nylas-base-950);cursor:pointer}";

const NylasNotification = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.ttl = 5000;
        this.allowedCategories = ['component', 'api', 'validation', 'invalid_session'];
        this.notifications = [];
    }
    onError(event) {
        utils.debug(`[nylas-notification] Error: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onWarning(event) {
        utils.debug(`[nylas-notification] Warning: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onInfo(event) {
        utils.debug(`[nylas-notification] Info: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onSuccess(event) {
        utils.debug(`[nylas-notification] Success: ${event.detail.title}`);
        this.handleNotification(event);
    }
    handleNotification(event) {
        if (event.defaultPrevented) {
            utils.debug(`[nylas-notification] Event default prevented`);
            return;
        }
        if (this.ttl === 0) {
            utils.debug(`[nylas-notification] TTL is 0`);
            return;
        }
        if (this.allowedCategories.length > 0 && !this.allowedCategories.includes(event.detail.category)) {
            utils.debug(`[nylas-notification] Category not allowed`);
            return;
        }
        this.notifications = [...this.notifications, { ...event.detail }];
        if (event.detail.ttl !== 'none') {
            const ttl = event.detail.ttl ? event.detail.ttl : this.ttl;
            window.setTimeout(() => {
                this.notifications = this.notifications.filter(notif => notif.id !== event.detail.id);
            }, ttl);
        }
    }
    dismissNotification(id) {
        this.notifications = this.notifications.filter(notif => notif.id !== id);
    }
    render() {
        return (index.h(index.Host, { key: '16bd673245819e3ed76ba5689c3292be510e3cd7' }, index.h(index.Fragment, { key: 'd88e8ea863d00a8c982c7e26007c975526c7457f' }, this.notifications.map(notif => (index.h("div", { class: `notif ${notif.type}`, key: notif.id }, index.h("div", { class: "content" }, index.h("span", { class: `icon ${notif.type}` }, (notif.type === 'error' || notif.type === 'warning') && index.h("warning-icon", null)), index.h("span", { class: "message" }, index.h("span", { class: "title" }, notif.title), notif.description && index.h("span", { class: "description" }, notif.description))), index.h("button", { class: "dismiss-btn", onClick: () => this.dismissNotification(notif.id) }, index.h("close-icon", null))))), index.h("slot", { key: 'c40cb61ce0a1dbd3416824b7c26e08e3aea01eae' }))));
    }
};
NylasNotification.style = nylasNotificationCss;

exports.nylas_notification = NylasNotification;
