// @nylas/web-elements/dist/cjs/nylas-list-folders.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c14ea8f5.js');
const registerComponent = require('./register-component-4f8ae42c.js');
const utils = require('./utils-85bd0909.js');
require('./_commonjsHelpers-ef8cd0cd.js');

const nylasListFoldersCss = ":host{display:block}.folders{display:flex;flex-direction:column;align-items:self-start}.folders .folder{border:1px solid var(--nylas-color-primary-100);padding:0.5rem 0.75rem;margin-bottom:0.5rem;justify-content:left;width:100%;min-width:200px;border-radius:var(--nylas-border-radius);background-color:var(--nylas-color-primary-50)}.folders .folder:last-child{margin-bottom:0}.folders .folder.selected{background-color:var(--nylas-color-primary-100)}.folders .folder .icon{margin-right:0.5rem;stroke:var(--nylas-color-primary-900);fill:none}hr{width:100%;margin:1rem 0;border:0;border-top:1px solid var(--nylas-color-primary-100)}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasListFolders = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.folderSelected = index.createEvent(this, "folderSelected", 7);
        this.folderSelectedHandler = (folder) => {
            this.folderSelected?.emit(folder);
        };
        this.folders = undefined;
        this.selectedFolder = undefined;
        this.defaultFolders = undefined;
        this.otherFolders = undefined;
    }
    connectedCallback() { }
    disconnectedCallback() { }
    async componentWillLoad() {
        utils.debug(`[nylas-list-folders] Component will load`);
    }
    async componentDidLoad() {
        utils.debug(`[nylas-list-folders] Component did load`);
    }
    watchFolders(newFolders) {
        utils.debug(`[nylas-list-folders] watchFolders`, { newFolders });
        const defaultFolders = ['inbox', 'drafts', 'sent', 'starred', 'spam', 'archive', 'trash'];
        this.defaultFolders = defaultFolders
            .filter(folderName => newFolders.find(folder => folder.name?.toLowerCase() === folderName.toLowerCase()))
            .map(folderName => newFolders.find(folder => folder.name?.toLowerCase() === folderName.toLowerCase()))
            .filter(folder => typeof folder !== 'undefined')
            .sort((a, b) => defaultFolders.indexOf(a?.name ?? '') - defaultFolders.indexOf(b?.name ?? ''));
        this.otherFolders = newFolders
            .filter(folder => !defaultFolders.find(folderName => folder.name?.toLowerCase() === folderName.toLowerCase()))
            .sort((a, b) => (a.name ?? '').localeCompare(b.name ?? ''));
    }
    formatFolderName(folder) {
        return utils.toTitleCase(folder.name?.replace('CATEGORY_', '') ?? 'Unknown');
    }
    getFolderIcon(folder) {
        switch (folder.name?.toLowerCase()) {
            case 'inbox':
                return index.h("inbox-icon", null);
            case 'drafts':
                return index.h("drafts-icon", null);
            case 'sent':
                return index.h("sent-icon", null);
            case 'starred':
                return index.h("star-icon", null);
            case 'spam':
                return index.h("spam-icon", null);
            case 'archive':
                return index.h("archive-icon", null);
            case 'trash':
                return index.h("trash-icon", null);
            default:
                return index.h("folder-icon", null);
        }
    }
    render() {
        return (index.h(index.Host, { key: '222ae81308d17cbdc7c3a44f0902fc683c88da1d' }, index.h("div", { key: '5ac1653c3b9ea1e3712e7597becc4d38435f3fa1', class: "folders" }, this.defaultFolders?.map(folder => (index.h("sp-button", { onClick: () => this.folderSelectedHandler(folder), class: {
                folder: true,
                selected: utils.formatFolderName(folder.name ?? 'inbox').toLowerCase() === this.selectedFolder?.toLowerCase(),
            } }, index.h("div", { slot: "icon", class: "icon" }, this.getFolderIcon(folder)), this.formatFolderName(folder))))), this.otherFolders && (index.h(index.Fragment, null, index.h("hr", null), index.h("div", { class: "folders" }, this.otherFolders?.map(folder => (index.h("sp-button", { onClick: () => this.folderSelectedHandler(folder), class: {
                folder: true,
                selected: utils.formatFolderName(folder.name ?? 'inbox').toLowerCase() === this.selectedFolder?.toLowerCase(),
            } }, index.h("div", { slot: "icon", class: "icon" }, this.getFolderIcon(folder)), this.formatFolderName(folder)))))))));
    }
    static get watchers() { return {
        "folders": ["watchFolders"]
    }; }
};
__decorate([
    registerComponent.RegisterComponent({
        name: 'nylas-pagination',
        stateToProps: new Map([
            ['mailbox.folders', 'folders'],
            ['mailbox.selectedFolder', 'selectedFolder'],
        ]),
        eventToProps: {
            folderSelected: async (_event, nylasConnector) => {
                utils.debug(`[nylas-list-folders] folderSelected event`, { _event });
                await nylasConnector.threads.getThreads({
                    page_token: undefined,
                    in: utils.formatFolderName(_event.detail.id ?? 'inbox'),
                });
            },
        },
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasListFolders.prototype, "render", null);
NylasListFolders.style = nylasListFoldersCss;

exports.nylas_list_folders = NylasListFolders;
