// @nylas/web-elements/dist/cjs/nylas-event-info.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c14ea8f5.js');
const registerComponent = require('./register-component-4f8ae42c.js');
const utils = require('./utils-85bd0909.js');
require('./_commonjsHelpers-ef8cd0cd.js');

const nylasEventInfoCss = ":host{display:block;margin:1rem;--nylas-primary:#2563eb;--nylas-error:#cc4841;--nylas-error-pressed:#992222;--nylas-wraning:#f06c00;--nylas-success:#16a392;--nylas-info:#2b8fc2;--nylas-base-0:#ffffff;--nylas-base-25:#fcfcfd;--nylas-base-50:#f8f9fc;--nylas-base-100:#eaecf5;--nylas-base-200:#d5d9eb;--nylas-base-300:#b3b8d8;--nylas-base-400:#717bbc;--nylas-base-500:#4e5ba6;--nylas-base-600:#3e4784;--nylas-base-700:#263f72;--nylas-base-800:#293056;--nylas-base-900:#101323;--nylas-base-950:#0e101b;--nylas-border-radius:0.25rem;--nylas-border-radius-2x:0.5rem;--nylas-border-radius-3x:0.75rem;--nylas-font-family:\"Inter\", sans-serif}.nylas-event-info__title{color:var(--nylas-base-900);font-size:16px;font-style:normal;font-weight:600;line-height:20px;margin:0;text-align:left}.nylas-event-info__subtitle{color:var(--nylas-base-600);font-size:14px;font-style:normal;font-weight:400;line-height:20px;margin:0.25rem 0px 0px;text-align:left}.nylas-event-info__content{padding:1rem;display:flex;flex-direction:column;gap:1rem}.nylas-event-info__block{display:flex;justify-content:space-between;gap:0.5rem}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasEventInfo = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
    }
    connectedCallback() {
        utils.debug('nylas-event-info', 'connectedCallback');
    }
    componentWillLoad() {
        utils.debug('nylas-event-info', 'componentWillLoad');
    }
    componentDidLoad() {
        utils.debug('nylas-event-info', 'componentDidLoad');
    }
    disconnectedCallback() {
        utils.debug('nylas-event-info', 'disconnectedCallback');
    }
    render() {
        return (index.h(index.Host, { key: 'e606a689e2b20dcddcbdbf42c44933953cb0c9c7' }, index.h("nylas-form-card", { key: 'f3827f8c8756d131ea0eeca39d8dcae783b58629' }, index.h("h1", { key: '6deb1da623b3c6ede66653201476647d30b8f7e5', slot: "header-title", class: "nylas-event-info__title" }, "Event information"), index.h("h3", { key: 'f027a8e56e59fba0e8ec3e4bfcbfda6419c5a6dc', slot: "header-subtitle", class: "nylas-event-info__subtitle" }, "Enter all event information."), index.h("div", { key: '0124599b210dc4e9395cdc75dd964132e98736b9', slot: "content", class: "nylas-event-info__content" }, index.h("slot", { key: 'a227128a7e6f7f8b8c52bd0bd24ce672979fef30', name: "inputs" })))));
    }
    static get formAssociated() { return true; }
};
__decorate([
    registerComponent.RegisterComponent({
        name: 'nylas-event-info',
        eventToProps: {},
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasEventInfo.prototype, "render", null);
NylasEventInfo.style = nylasEventInfoCss;

exports.nylas_event_info = NylasEventInfo;
