// @spectrum-web-components/dialog/stories/index.js
"use strict";
import { html } from "@spectrum-web-components/base";
function nextFrame() {
  return new Promise((res) => requestAnimationFrame(() => res()));
}
class OverlayTriggerReady extends HTMLElement {
  constructor() {
    super(...arguments);
    this.handleTriggerOpened = async () => {
      await nextFrame();
      this.ready(true);
    };
    this.readyPromise = Promise.resolve(false);
  }
  connectedCallback() {
    this.readyPromise = new Promise((res) => {
      this.ready = res;
      this.setup();
    });
  }
  async setup() {
    await nextFrame();
    await nextFrame();
    const overlay = document.querySelector(
      `overlay-trigger[open]`
    );
    overlay.addEventListener("sp-opened", this.handleTriggerOpened);
  }
  get updateComplete() {
    return this.readyPromise;
  }
}
customElements.define("overlay-trigger-ready", OverlayTriggerReady);
export const overlayTriggerDecorator = (story) => {
  return html`
        ${story()}
        <overlay-trigger-ready></overlay-trigger-ready>
    `;
};
class CountdownWatcher extends HTMLElement {
  constructor() {
    super(...arguments);
    this.readyPromise = Promise.resolve(false);
  }
  connectedCallback() {
    this.previousElementSibling.addEventListener(
      "countdown-complete",
      () => {
        this.ready(true);
      }
    );
    this.readyPromise = new Promise((res) => {
      this.ready = res;
    });
  }
  get updateComplete() {
    return this.readyPromise;
  }
}
customElements.define("countdown-complete-watcher", CountdownWatcher);
export const disabledButtonDecorator = (story) => {
  return html`
        ${story()}
        <countdown-complete-watcher></countdown-complete-watcher>
    `;
};
