// @nylas/web-elements/dist/collection/stores/mailbox-store.js
import { createStore } from "@stencil/store";
export function CreateNylasMailboxStore(defaultState = {}) {
    const defaultNylasStoreState = {
        state: 'ready',
        messages: [],
        threads: [],
        folders: [],
        selectedThread: null,
        nextThreadPageToken: null,
        previousThreadPageTokens: [],
        searchQuery: '',
        hasNextThreadPage: false,
        hasPreviousThreadPage: false,
        selectedFolder: 'inbox',
        threadsQueryParams: {
            in: 'inbox',
        },
        messgeQueryParams: {
            fields: 'include_headers',
        },
        ...defaultState,
    };
    const store = createStore({ ...defaultNylasStoreState });
    store.onChange('nextThreadPageToken', nextThreadPageToken => {
        store.set('hasNextThreadPage', nextThreadPageToken !== null);
    });
    store.onChange('previousThreadPageTokens', previousThreadPageTokens => {
        store.set('hasPreviousThreadPage', previousThreadPageTokens.length > 0);
    });
    store.onChange('threadsQueryParams', threadsQueryParams => {
        store.set('selectedFolder', threadsQueryParams.in || 'inbox');
        store.set('searchQuery', threadsQueryParams.search_query_native || '');
    });
    store.reset = () => {
        for (const key in defaultNylasStoreState) {
            const value = defaultNylasStoreState[key];
            store.set(key, value);
        }
    };
    return store;
}
