// @nylas/web-elements/dist/collection/connector/nylas-scheduler-config-connector/index.js
import { debug } from "../../utils/utils";
import { NylasSchedulerConfigAPIConnector } from "../shared/api/scheduler-config";
export class NylasSchedulerConfigConnector {
    constructor({ nylasApiRequest, nylasSchedulerConfigStore }) {
        this.nylasApiRequest = nylasApiRequest;
        this.nylasSchedulerConfigStore = nylasSchedulerConfigStore;
    }
    init() {
        debug('NylasSchedulerConfigConnector init');
    }
    get schedulerConfigStore() {
        return this.nylasSchedulerConfigStore;
    }
    get schedulerConfig() {
        if (this._schedulerConfigConnector) {
            return this._schedulerConfigConnector;
        }
        this._schedulerConfigConnector = new NylasSchedulerConfigAPIConnector({
            nylasApiRequest: this.nylasApiRequest,
            nylasSchedulerConfigStore: this.nylasSchedulerConfigStore,
        });
        return this._schedulerConfigConnector;
    }
    resetStores() {
        debug('NylasSchedulerConfigConnector resetStores');
        this.nylasSchedulerConfigStore.reset();
    }
    async logout() {
        debug('NylasSchedulerConfigConnector logout');
        this.resetStores();
    }
}
