// @nylas/web-elements/dist/collection/components/nylas-provider/nylas-provider.js
import { NylasBaseProvider } from "../../common/abstract-provider";
import { NylasConnector } from "../../connector/nylas-connector/index";
import { HashRouter } from "../../routers/hash-router";
import { CreateNylasAuthStore } from "../../stores/auth-store";
import { CreateNylasMailboxStore } from "../../stores/mailbox-store";
import { CreateNylasSchedulerStore } from "../../stores/scheduler-store";
import { NylasAuth } from "@nylas/core";
import { Host, h } from "@stencil/core";
import { debug } from "../../utils/utils";
import { CreateNylasSchedulerConfigStore } from "../../stores/scheduler-config-store";
export class NylasProvider {
    constructor() {
        this.handleAuthChange = async (isAuthenticated) => {
            if (isAuthenticated) {
                this.loggedIn.emit(this.host);
            }
            else {
                this.loggedOut.emit(this.host);
            }
        };
        this.authConfig = undefined;
        this.eventOverrides = {};
        this.automaticComponentRegistration = true;
    }
    connectedCallback() {
        debug('[nylas-provider] connectedCallback');
    }
    async componentWillLoad() {
        debug('[nylas-provider] componentWillLoad');
        this.stores = {
            auth: CreateNylasAuthStore(),
            mailbox: CreateNylasMailboxStore(),
            scheduler: CreateNylasSchedulerStore(),
            schedulerConfig: CreateNylasSchedulerConfigStore(),
        };
        this.baseProvider = new NylasBaseProvider(this.host, this.stores, this.automaticComponentRegistration, this.eventOverrides);
        const router = new HashRouter();
        this.nylasAuth = NylasAuth({
            apiUri: 'https://api.us.nylas.com',
            elementsApiUri: 'https://elements.us.nylas.com',
            storageType: 'indexeddb',
            clientId: 'YOUR_CLIENT_ID',
            defaultScopes: [],
            redirectURI: '/',
            ...this.authConfig,
        });
        const nylasMailboxStore = this.baseProvider?.getStore('mailbox');
        if (!nylasMailboxStore) {
            throw new Error('The mailbox store is not set');
        }
        const nylasAuthStore = this.baseProvider?.getStore('auth');
        if (!nylasAuthStore) {
            throw new Error('The auth store is not set');
        }
        const nylasSchedulerStore = this.baseProvider?.getStore('scheduler');
        if (!nylasSchedulerStore) {
            throw new Error('The scheduler store is not set');
        }
        const nylasSchedulerConfigStore = this.baseProvider?.getStore('schedulerConfig');
        if (!nylasSchedulerConfigStore) {
            throw new Error('The scheduler config store is not set');
        }
        this.nylasConnector = new NylasConnector(router, this.nylasAuth, nylasAuthStore, nylasMailboxStore, nylasSchedulerStore, nylasSchedulerConfigStore);
        const initEvent = this.init.emit(this.host);
        if (!initEvent.defaultPrevented) {
            await this.nylasConnector?.auth.validateSession();
        }
        this.handleAuthChange(nylasAuthStore.state.isAuthenticated);
        this.baseProvider?.componentWillLoad(this.nylasConnector);
    }
    componentDidLoad() {
        const nylasAuthStore = this.baseProvider?.getStore('auth');
        nylasAuthStore?.onChange('isAuthenticated', this.handleAuthChange);
        this.baseProvider?.componentDidLoad();
    }
    componentDisconnected() {
        this.baseProvider?.componentDisconnected();
    }
    async registerComponentHandler(event) {
        this.baseProvider?.registerComponent(event.detail);
    }
    async unregisterComponentHandler(event) {
        this.baseProvider?.unregisterComponent(event.detail);
    }
    async getNylasAuth() {
        return this.nylasAuth;
    }
    async getNylasConnector() {
        return this.nylasConnector;
    }
    async getMailboxStore() {
        return this.baseProvider?.getStore('mailbox');
    }
    async getAuthStore() {
        return this.baseProvider?.getStore('auth');
    }
    async getNylasSchedulerStore() {
        return this.baseProvider?.getStore('scheduler');
    }
    async getNylasSchedulerConfigStore() {
        return this.baseProvider?.getStore('schedulerConfig');
    }
    render() {
        return (h(Host, { key: '782d47608825fb870d87153c8d43d778dc6c2b87' }, h("slot", { key: '2b01537d9d4a0cf4c81ad0b77ce3e6d761f3e1db' })));
    }
    static get is() { return "nylas-provider"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["nylas-provider.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["nylas-provider.css"]
        };
    }
    static get properties() {
        return {
            "authConfig": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "AuthConfig",
                    "resolved": "undefined | { apiUri: string; elementsApiUri: string; clientId: string; defaultScopes: string[]; storageType: \"indexeddb\" | \"localstorage\"; redirectURI: string; }",
                    "references": {
                        "AuthConfig": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::AuthConfig"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The Nylas Auth configuration.\nUsed to manage all things authentication with Nylas."
                }
            },
            "eventOverrides": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "EventOverride<Exclude<typeof this.nylasConnector, undefined>>",
                    "resolved": "{ [x: string]: (event: CustomEvent<any>, connector?: NylasConnector | undefined) => Promise<void>; }",
                    "references": {
                        "EventOverride": {
                            "location": "import",
                            "path": "@/common/component-types",
                            "id": "src/common/component-types.ts::EventOverride"
                        },
                        "Exclude": {
                            "location": "global",
                            "id": "global::Exclude"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "This provides a way to override the default event handlers."
                },
                "defaultValue": "{}"
            },
            "automaticComponentRegistration": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "true"
                        }],
                    "text": "Automatically register components that have the `@RegisterComponent` decorator.\nIf this is set to false, you will need to manually register components using the\n`registerComponent` method."
                },
                "attribute": "automatic-component-registration",
                "reflect": false,
                "defaultValue": "true"
            }
        };
    }
    static get events() {
        return [{
                "method": "init",
                "name": "init",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the provider is initialized.\nIt can be used to set the initial state of the provider,\nor to prevent the provider from firing some default behavior."
                },
                "complexType": {
                    "original": "HTMLNylasProviderElement",
                    "resolved": "HTMLNylasProviderElement",
                    "references": {
                        "HTMLNylasProviderElement": {
                            "location": "global",
                            "id": "global::HTMLNylasProviderElement"
                        }
                    }
                }
            }, {
                "method": "loggedIn",
                "name": "loggedIn",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the the NyalsAuth isAuthenticated state changes\nto true."
                },
                "complexType": {
                    "original": "HTMLNylasProviderElement",
                    "resolved": "HTMLNylasProviderElement",
                    "references": {
                        "HTMLNylasProviderElement": {
                            "location": "global",
                            "id": "global::HTMLNylasProviderElement"
                        }
                    }
                }
            }, {
                "method": "loggedOut",
                "name": "loggedOut",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "This event is fired when the the NyalsAuth isAuthenticated state changes\nto true."
                },
                "complexType": {
                    "original": "HTMLNylasProviderElement",
                    "resolved": "HTMLNylasProviderElement",
                    "references": {
                        "HTMLNylasProviderElement": {
                            "location": "global",
                            "id": "global::HTMLNylasProviderElement"
                        }
                    }
                }
            }];
    }
    static get methods() {
        return {
            "getNylasAuth": {
                "complexType": {
                    "signature": "() => Promise<NylasAuthType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "AuthConfig": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::AuthConfig"
                        },
                        "RequestInit": {
                            "location": "global",
                            "id": "global::RequestInit"
                        },
                        "T": {
                            "location": "global",
                            "id": "global::T"
                        },
                        "ReadableStream": {
                            "location": "global",
                            "id": "global::ReadableStream"
                        },
                        "Uint8Array": {
                            "location": "global",
                            "id": "global::Uint8Array"
                        },
                        "T_2": {
                            "location": "global",
                            "id": "global::T_2"
                        },
                        "T_3": {
                            "location": "global",
                            "id": "global::T_3"
                        },
                        "Blob": {
                            "location": "global",
                            "id": "global::Blob"
                        },
                        "NylasAuthType": {
                            "location": "import",
                            "path": "@nylas/core",
                            "id": "../nylas-js-core/dist/index.d.ts::NylasAuthType"
                        }
                    },
                    "return": "Promise<{ config: AuthConfig; getHostedAuthRedirectURL(scopes?: string[] | undefined, email?: string | undefined, provider?: string | undefined): Promise<string>; nylasAPIRequest: { <T>(path: string, method?: string | undefined, body?: any, parseJSON?: boolean | undefined, domain?: string | undefined, request?: RequestInit | undefined): Promise<T>; <T_1>(path: string, method: string, body: any, parseJSON: false, domain?: string | undefined, request?: RequestInit | undefined): Promise<ReadableStream<Uint8Array> | null>; <T_2>(path: string, method: string, body: any, parseJSON: true, domain?: string | undefined, request?: RequestInit | undefined): Promise<T_2>; <T_3>(path: string, method: string, body: any, parseJSON: boolean, domain?: string | undefined, request?: RequestInit | undefined): Promise<T_3 | ReadableStream<Uint8Array> | null>; }; exchangeAuthCodeForTokenInfo(code: string): Promise<TokenExchangeResponse | null>; nylasDownloadFileRequest(path: string, method?: string | undefined, body?: any): Promise<Blob>; logout(): Promise<void>; isAuthenticated(): Promise<boolean>; getTokenInfo(): Promise<TokenInfo | null>; isAccessTokenValid(): Promise<boolean>; refreshAccessToken(): Promise<TokenExchangeResponse | null>; hasScopeByType(service: \"email\" | undefined, types: string | string[]): Promise<boolean>; hasScope(scope: string | string[]): Promise<boolean>; } | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasAuth instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasAuth instance"
                        }]
                }
            },
            "getNylasConnector": {
                "complexType": {
                    "signature": "() => Promise<NylasConnector | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasConnector": {
                            "location": "import",
                            "path": "@/connector/nylas-connector",
                            "id": "src/connector/nylas-connector/index.ts::NylasConnector"
                        }
                    },
                    "return": "Promise<NylasConnector | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasConnector instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasConnector instance"
                        }]
                }
            },
            "getMailboxStore": {
                "complexType": {
                    "signature": "() => Promise<NylasMailboxStoreType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasMailboxStoreType": {
                            "location": "import",
                            "path": "@/stores/mailbox-store",
                            "id": "src/stores/mailbox-store.ts::NylasMailboxStoreType"
                        }
                    },
                    "return": "Promise<ObservableMap<NylasMailboxStoreState> | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the mailboxStore instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The mailboxStore instance"
                        }]
                }
            },
            "getAuthStore": {
                "complexType": {
                    "signature": "() => Promise<NylasAuthStoreType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasAuthStoreType": {
                            "location": "import",
                            "path": "@/stores/auth-store",
                            "id": "src/stores/auth-store.ts::NylasAuthStoreType"
                        }
                    },
                    "return": "Promise<ObservableMap<NylasAuthStoreState> | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the authStore instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The authStore instance"
                        }]
                }
            },
            "getNylasSchedulerStore": {
                "complexType": {
                    "signature": "() => Promise<NylasSchedulerStoreType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasSchedulerStoreType": {
                            "location": "import",
                            "path": "@/stores/scheduler-store",
                            "id": "src/stores/scheduler-store.ts::NylasSchedulerStoreType"
                        }
                    },
                    "return": "Promise<ObservableMap<NylasSchedulerStoreState> | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasScheduler instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasScheduler instance"
                        }]
                }
            },
            "getNylasSchedulerConfigStore": {
                "complexType": {
                    "signature": "() => Promise<NylasSchedulerConfigStoreType | undefined>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "NylasSchedulerConfigStoreType": {
                            "location": "import",
                            "path": "@/stores/scheduler-config-store",
                            "id": "src/stores/scheduler-config-store.ts::NylasSchedulerConfigStoreType"
                        }
                    },
                    "return": "Promise<CreateNylasSchedulerConfigStoreReturnType | undefined>"
                },
                "docs": {
                    "text": "This method is used to retrieve the NylasSchedulerConfig instance",
                    "tags": [{
                            "name": "returns",
                            "text": "The NylasSchedulerConfig instance"
                        }]
                }
            }
        };
    }
    static get elementRef() { return "host"; }
    static get listeners() {
        return [{
                "name": "registerComponent",
                "method": "registerComponentHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "unregisterComponent",
                "method": "unregisterComponentHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
