// @nylas/web-elements/dist/components/button-component2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const buttonComponentCss = "button.sc-button-component{height:48px;border:none;border-radius:var(--nylas-border-radius-2x);font-size:1rem;line-height:24px;padding:0 1rem;cursor:pointer;font-family:var(--nylas-font-family);font-weight:600}@media screen and (max-width: 768px){button.sc-button-component{width:100%}}button.button-primary.sc-button-component{background:var(--nylas-primary);color:var(--nylas-base-0);height:48px;border:none;border-radius:var(--nylas-border-radius-2x);font-size:1rem;line-height:24px;padding:0 1rem;cursor:pointer}button.button-primary.sc-button-component:hover{background:var(--nylas-base-600)}button.button-primary.sc-button-component:active{background:var(--nylas-base-800)}button.button-primary.sc-button-component:disabled{background:var(--nylas-base-300);cursor:not-allowed}button.button-basic.sc-button-component{border:1px solid var(--nylas-base-300);color:var(--nylas-base-700);background:transparent}button.button-basic.sc-button-component:hover{background:transparent;border-color:var(--nylas-primary)}button.button-basic.sc-button-component:active{background:transparent;border-color:var(--nylas-base-600);border-width:2px}button.button-basic.sc-button-component:disabled{background:transparent;border-color:var(--nylas-base-300);color:var(--nylas-base-300);cursor:not-allowed}button.button-destructive.sc-button-component{border:1px solid var(--nylas-base-300);color:var(--nylas-error);background:transparent}button.button-destructive.sc-button-component:hover{background:transparent;border-color:var(--nylas-error)}button.button-destructive.sc-button-component:active{background:transparent;border-color:var(--nylas-error-pressed);border-width:2px;color:var(--nylas-error-pressed)}button.button-destructive.sc-button-component:disabled{background:transparent;border-color:var(--nylas-base-300);color:var(--nylas-base-300);cursor:not-allowed}";

const ButtonComponent = proxyCustomElement(class ButtonComponent extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.variant = 'primary';
        this.disabled = false;
        this.type = 'button';
        this.clickHandler = undefined;
        this.mouseOverHandler = undefined;
        this.mouseOutHandler = undefined;
        this.focusHandler = undefined;
        this.blurHandler = undefined;
    }
    render() {
        const className = `button-${this.variant}` + (this.disabled ? ' disabled' : '');
        return (h("button", { key: '3ac840e8b65854c2e4bdabe93b21b5bcda99e4db', class: className, disabled: this.disabled, type: "type", onClick: this.clickHandler, onMouseOver: this.mouseOverHandler, onMouseOut: this.mouseOutHandler, onFocus: this.focusHandler, onBlur: this.blurHandler }, h("slot", { key: '110df0711566be0df89f2e04aed008dcaf91e9f3' })));
    }
    static get style() { return buttonComponentCss; }
}, [6, "button-component", {
        "variant": [1],
        "disabled": [4],
        "type": [1],
        "clickHandler": [16],
        "mouseOverHandler": [16],
        "mouseOutHandler": [16],
        "focusHandler": [16],
        "blurHandler": [16]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["button-component"];
    components.forEach(tagName => { switch (tagName) {
        case "button-component":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ButtonComponent);
            }
            break;
    } });
}

export { ButtonComponent as B, defineCustomElement as d };
