// @nylas/web-elements/dist/components/calendar-cancel.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-calendar-cancel-icon-h{display:flex}";

const CalendarCancelIcon = proxyCustomElement(class CalendarCancelIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '20';
        this.height = '20';
    }
    render() {
        return (h("svg", { key: '09bdd81a98611d49143ed327802cee77c1e851ae', width: this.width, height: this.height, fill: "none", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '9692e20de89c490c05ccf9800faea78c5860260d', d: "M14.5 9a5.5 5.5 0 1 1 0 11 5.5 5.5 0 0 1 0-11Zm-2.476 3.024a.5.5 0 0 0 0 .707l1.769 1.77-1.767 1.766a.5.5 0 1 0 .707.708l1.767-1.767 1.77 1.769a.5.5 0 1 0 .707-.707L15.208 14.5l1.771-1.77a.5.5 0 0 0-.707-.707l-1.771 1.77-1.77-1.77a.5.5 0 0 0-.707 0ZM14.75 0A3.25 3.25 0 0 1 18 3.25l.001 5.773a6.47 6.47 0 0 0-1.5-.71L16.5 5.5h-15v9.25c0 .966.784 1.75 1.75 1.75h5.064a6.47 6.47 0 0 0 .709 1.501L3.25 18A3.25 3.25 0 0 1 0 14.75V3.25A3.25 3.25 0 0 1 3.25 0h11.5Zm0 1.5H3.25A1.75 1.75 0 0 0 1.5 3.25V4h15v-.75a1.75 1.75 0 0 0-1.75-1.75Z", fill: "currentColor" })));
    }
    static get style() { return iconCss; }
}, [2, "calendar-cancel-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["calendar-cancel-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "calendar-cancel-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CalendarCancelIcon);
            }
            break;
    } });
}

export { CalendarCancelIcon as C, defineCustomElement as d };
