// @nylas/web-elements/dist/components/checkmark-circle.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-checkmark-circle-icon-h{display:flex}";

const CheckmarkCircleIcon = proxyCustomElement(class CheckmarkCircleIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '20';
        this.height = '20';
    }
    render() {
        return (h("svg", { key: '391cc64dd53e54e877a74042d20890c59b1a25a1', xmlns: "http://www.w3.org/2000/svg", width: this.width, height: this.height, fill: "none", viewBox: "0 0 20 20" }, h("path", { key: '9fa7b6dd7971961e6129720a4ec73af505d5c188', fill: "currentColor", d: "M10 2a8 8 0 1 1 0 16 8 8 0 0 1 0-16Zm0 1a7 7 0 1 0 0 14 7 7 0 0 0 0-14Zm3.358 4.646a.5.5 0 0 1 .058.638l-.058.07-4.004 4.004a.5.5 0 0 1-.638.058l-.07-.058-2-2a.5.5 0 0 1 .638-.765l.07.058L9 11.298l3.651-3.652a.5.5 0 0 1 .707 0Z" })));
    }
    static get style() { return iconCss; }
}, [2, "checkmark-circle-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["checkmark-circle-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "checkmark-circle-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CheckmarkCircleIcon);
            }
            break;
    } });
}

export { CheckmarkCircleIcon as C, defineCustomElement as d };
