// @nylas/web-elements/dist/components/chevron.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-chevron-icon-h{display:flex}";

const ChevronIcon = proxyCustomElement(class ChevronIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '24';
        this.height = '24';
    }
    render() {
        return (h("svg", { key: '4237d5be45ec5b20c76f6132347abbbff3db30d3', xmlns: "http://www.w3.org/2000/svg", width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "none" }, h("path", { key: '3d68979af03e88ecb6725fefbd1a4d9af22aa727', fill: "currentColor", d: "M15.53 4.22a.75.75 0 0 1 0 1.06L8.81 12l6.72 6.72a.75.75 0 1 1-1.06 1.06l-7.25-7.25a.75.75 0 0 1 0-1.06l7.25-7.25a.75.75 0 0 1 1.06 0Z" })));
    }
    static get style() { return iconCss; }
}, [2, "chevron-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["chevron-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "chevron-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ChevronIcon);
            }
            break;
    } });
}

export { ChevronIcon as C, defineCustomElement as d };
