// @nylas/web-elements/dist/components/close.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-close-icon-h{display:flex}";

const CloseIcon = proxyCustomElement(class CloseIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '16';
        this.height = '16';
    }
    render() {
        return (h("svg", { key: 'e58d42880affc512eb7ae18a34099340a50dc6be', width: "21", height: "21", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, h("g", { key: '122df83308958e375531cdaa9d057b78d36b08e3', "clip-path": "url(#a)" }, h("path", { key: '9b712ee4fa93a86f9a5609bea9233f17160bc23e', d: "M5.73 4.574a.75.75 0 1 0-1.061 1.06l4.419 4.42-4.42 4.42a.75.75 0 1 0 1.061 1.06l4.42-4.42 4.419 4.42a.75.75 0 0 0 1.06-1.06l-4.419-4.42 4.42-4.42a.75.75 0 1 0-1.061-1.06l-4.42 4.42-4.419-4.42Z", fill: "currentColor" })), h("defs", { key: '761a14ba84069f692cad52f6eafa8cafc4a36489' }, h("clipPath", { key: '6805481c67c08d9253c6a5bd7ece2688f6fe0e55', id: "a" }, h("path", { key: 'af3d03843001cd7418021a9dd216f2e2f126fc7f', fill: "#fff", transform: "translate(.148 .054)", d: "M0 0h20v20H0z" })))));
    }
    static get style() { return iconCss; }
}, [2, "close-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["close-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "close-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CloseIcon);
            }
            break;
    } });
}

export { CloseIcon as C, defineCustomElement as d };
