// @nylas/web-elements/dist/components/inbox.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-inbox-icon-h{display:flex}";

const InboxIcon = proxyCustomElement(class InboxIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("svg", { key: '25cc88e340b2a67cd5bb5cfd251389f879315479', width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "auto", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '90e4e7c4e69f67ec0721f3cac5be85898d0b3188', d: "M20 3H4C2.89543 3 2 3.89543 2 5V19C2 20.1046 2.89543 21 4 21H20C21.1046 21 22 20.1046 22 19V5C22 3.89543 21.1046 3 20 3Z", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: '458347d7c012ac02b040b1b374c447c4addadf3e', d: "M2 5L12 12.5L22 5", stroke: "auto", "stroke-width": "1.5" })));
    }
    static get style() { return iconCss; }
}, [2, "inbox-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["inbox-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "inbox-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, InboxIcon);
            }
            break;
    } });
}

export { InboxIcon as I, defineCustomElement as d };
