// @nylas/web-elements/dist/components/loading.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const loadingCss = ".wave{display:flex;align-items:center;justify-content:center}.dot{width:4px;height:4px;background-color:var(--dot-color, white);border-radius:50%;margin:0 2px;animation:wave 1.5s infinite ease-in-out}.dot:nth-child(2){animation-delay:0.2s}.dot:nth-child(3){animation-delay:0.4s}@keyframes wave{0%,60%,100%{transform:translateY(0)}30%{transform:translateY(-4px)}}";

const LoadingIcon = proxyCustomElement(class LoadingIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("div", { key: '20e081b4657135796ba486fe3b3ba4eb7e24df55', class: "wave" }, h("div", { key: 'ba8f7cd30e8dbdc4ca5fae882662563661948493', class: "dot" }), h("div", { key: '57edfc7797ee57ab9ae22fef20a5abcb84c3bc6b', class: "dot" }), h("div", { key: 'c56c73301414ebaf6b66112dd3745d7e72f11508', class: "dot" })));
    }
    static get style() { return loadingCss; }
}, [1, "loading-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["loading-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "loading-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, LoadingIcon);
            }
            break;
    } });
}

export { LoadingIcon as L, defineCustomElement as d };
