// @nylas/web-elements/dist/components/microsoft-logo.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-microsoft-logo-icon-h{display:flex}";

const MicrosoftLogoIcon = proxyCustomElement(class MicrosoftLogoIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '22';
        this.height = '22';
    }
    render() {
        return (h("svg", { key: 'de39bfb3167fbd60d21025e74c1bc19cdd7d4dc9', width: this.width, height: this.height, viewBox: "0 0 22 22", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '0f80864e892ff36e6e76ef03f5f36d62007d99ee', d: "M0.043457 0.043457H10.4782V10.4782H0.043457V0.043457Z", fill: "#F35325" }), h("path", { key: 'ab0e85f803a3cf2e32055ebeff1b0705edebe048', d: "M11.5217 0.043457H21.9565V10.4782H11.5217V0.043457Z", fill: "#81BC06" }), h("path", { key: 'b6359dcb76ade260f2a9e9bdf1be21ba84b77b79', d: "M0.043457 11.5217H10.4782V21.9565H0.043457V11.5217Z", fill: "#05A6F0" }), h("path", { key: 'cf38a555c09aaef94a15fd26b0951e2642eaad6d', d: "M11.5217 11.5217H21.9565V21.9565H11.5217V11.5217Z", fill: "#FFBA08" })));
    }
    static get style() { return iconCss; }
}, [2, "microsoft-logo-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["microsoft-logo-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "microsoft-logo-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MicrosoftLogoIcon);
            }
            break;
    } });
}

export { MicrosoftLogoIcon as M, defineCustomElement as d };
