// @nylas/web-elements/dist/components/multi-select-dropdown2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { a as debug } from './utils.js';
import { d as defineCustomElement$2 } from './chevron.js';
import { d as defineCustomElement$1 } from './close.js';

const multiSelectDropdownCss = ":host{display:block;width:inherit;--nylas-primary:#2563eb;--nylas-error:#cc4841;--nylas-error-pressed:#992222;--nylas-wraning:#f06c00;--nylas-success:#16a392;--nylas-info:#2b8fc2;--nylas-base-0:#ffffff;--nylas-base-25:#fcfcfd;--nylas-base-50:#f8f9fc;--nylas-base-100:#eaecf5;--nylas-base-200:#d5d9eb;--nylas-base-300:#b3b8d8;--nylas-base-400:#717bbc;--nylas-base-500:#4e5ba6;--nylas-base-600:#3e4784;--nylas-base-700:#263f72;--nylas-base-800:#293056;--nylas-base-900:#101323;--nylas-base-950:#0e101b;--nylas-border-radius:0.25rem;--nylas-border-radius-2x:0.5rem;--nylas-border-radius-3x:0.75rem;--nylas-font-family:\"Inter\", sans-serif}.dropdown{display:inline-block;width:100%;position:relative}.dropdown .dropdown-label{display:flex;align-items:center;gap:0.25rem;color:var(--nylas-base-800)}.dropbtn{width:inherit;height:48px;color:black;padding:0.5rem;font-size:1rem;cursor:pointer;display:flex;gap:0.5rem;justify-content:space-between;align-items:center;background:transparent;border:1px solid var(--nylas-base-200);border-radius:var(--nylas-border-radius-2x)}.dropbtn.focus{background:transparent}.dropbtn:hover,.dropbtn:active{border:1px solid var(--nylas-primary)}.dropbtn:active{outline:2px solid var(--nylas-primary)}.dropbtn span.open{transform:rotate(90deg)}.dropbtn span.closed{transform:rotate(270deg)}.dropdown-content{display:block;margin-top:0.5rem;background-color:var(--nylas-base-0);width:100%;max-height:336px;overflow:auto;border:1px solid #ddd;z-index:1;border-radius:4px;position:absolute;box-shadow:0px 4px 6px -2px rgba(0, 0, 0, 0.0509803922);box-shadow:0px 10px 15px -3px rgba(0, 0, 0, 0.1019607843)}.dropdown-content ul{padding:0;list-style-type:none;color:var(--nylas-base-900);max-height:336px}.dropdown-content ul li{padding:16px, 12px, 16px, 12px;color:black;padding:12px 16px;text-decoration:none;display:block;font-family:inherit;font-size:14px;font-weight:400;line-height:20px;letter-spacing:0px;text-align:left;cursor:pointer}.dropdown-content ul li:hover,.dropdown-content ul li:focus{background-color:var(--nylas-base-100)}.dropdown-content ul li label{display:flex;align-items:center;gap:0.5rem}.dropdown-content ul li label input{margin:0}.selected-options{display:flex;flex-wrap:wrap;gap:0.5rem;padding:0.5rem 0;margin-top:0.25rem;background:var(--nylas-base-0)}.selected-options .selected-option{display:flex;align-items:center;gap:0.5rem;padding:4px 8px;border-radius:var(--nylas-border-radius-2x);background:var(--nylas-base-100);color:var(--nylas-base-800);font-size:16px;font-weight:500;line-height:1.5rem;letter-spacing:0.5px}.selected-options .selected-option button{background:transparent;border:none;cursor:pointer;padding:0}.selected-options .selected-option button:hover{color:var(--nylas-primary)}";

const MultiSelectDropdown = proxyCustomElement(class MultiSelectDropdown extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.selectedOptionsChanged = createEvent(this, "selectedOptionsChanged", 7);
        this.name = undefined;
        this.label = undefined;
        this.options = [];
        this.availableOptions = this.options;
        this.isOpen = false;
        this.ariaActivedescendant = '';
        this.shouldFocusFirstOption = false;
    }
    componentDidRender() {
        debug('multi-select-dropdown', 'componentDidRender');
        if (this.isOpen && this.shouldFocusFirstOption) {
            this.ariaActivedescendant = this.availableOptions[0].value;
            this.focusOption(0);
            this.shouldFocusFirstOption = false;
        }
    }
    handleOutsideClick(event) {
        const path = event.composedPath();
        const isClickInside = path.includes(this.el);
        if (!isClickInside && this.isOpen) {
            this.isOpen = false;
        }
    }
    selectOption(option) {
        this.availableOptions = this.availableOptions.map(o => {
            if (o.value === option.value) {
                o.selected = option.selected ? false : true;
            }
            return o;
        });
        const selectedOptions = this.availableOptions.filter(o => o.selected).map(o => o.value);
        this.selectedOptionsChanged.emit({
            value: selectedOptions,
            name: this.name,
        });
    }
    toggleDropdown() {
        this.isOpen = !this.isOpen;
        if (this.isOpen) {
            this.shouldFocusFirstOption = true;
        }
        else {
            this.ariaActivedescendant = '';
        }
    }
    handleSelectButtonKeyDown(event) {
        switch (event.key) {
            case 'ArrowDown':
            case 'Enter':
                event.preventDefault();
                if (!this.isOpen) {
                    this.toggleDropdown();
                }
                break;
            case 'Escape':
                this.isOpen = false;
                break;
        }
    }
    handleListboxKeydown(e) {
        const items = this.availableOptions;
        const currentIndex = items.findIndex(item => item.value === this.ariaActivedescendant);
        switch (e.key) {
            case 'ArrowDown':
            case 'Tab':
                if (!e.shiftKey) {
                    e.preventDefault();
                    const nextIndex = currentIndex + 1 < items.length ? currentIndex + 1 : 0;
                    this.ariaActivedescendant = items[nextIndex].value;
                    this.focusOption(nextIndex);
                }
                else {
                    e.preventDefault();
                    const prevIndex = currentIndex - 1 >= 0 ? currentIndex - 1 : items.length - 1;
                    this.ariaActivedescendant = items[prevIndex].value;
                    this.focusOption(prevIndex);
                }
                break;
            case 'ArrowUp':
                e.preventDefault();
                const prevIndex = currentIndex - 1 >= 0 ? currentIndex - 1 : items.length - 1;
                this.ariaActivedescendant = items[prevIndex].value;
                this.focusOption(prevIndex);
                break;
            case 'Enter':
                e.preventDefault();
                if (this.ariaActivedescendant) {
                    this.selectOption(items[currentIndex]);
                }
                break;
            case 'Escape':
                this.isOpen = false;
                break;
        }
    }
    focusOption(index) {
        const option = this.availableOptions[index];
        if (!option)
            return;
        const elementId = option.value;
        const element = this.el.shadowRoot?.getElementById(elementId);
        if (element) {
            element.focus();
            element.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
        }
    }
    getSelectedOptions() {
        return this.availableOptions.filter(option => option.selected);
    }
    renderOption(option) {
        return (h("li", { key: option.value, id: option.value, role: "option", tabindex: "0", "aria-selected": option.selected ? 'true' : 'false', onClick: e => {
                e.stopImmediatePropagation();
                this.selectOption(option);
            }, class: { selected: !!option.selected } }, h("label", { htmlFor: option.value }, h("input", { "aria-hidden": "true", id: option.value, type: "checkbox", checked: option.selected }), h("span", null, option.label))));
    }
    render() {
        return (h(Host, { key: '4ef13670e3dfbf55e023ac7532d93c48d21f3004' }, h("div", { key: '8626f138f5be677421697067460ef76d0a7cc697', class: "dropdown", part: "msd_dropdown" }, h("label", { key: 'b723c488dac91c39f285aae1a3f7928844753bf1', class: "dropdown-label" }, this.label, h("slot", { key: '5bdec320647a41de57671c75523a453ccb07af2a', name: "label-icon", "aria-hidden": "true" })), h("button", { key: 'f7d7cf144ac2fec93171b5ac6cd42be4be6b5c93', part: "msd_dropdown-button", class: { dropbtn: true, open: this.isOpen }, onClick: () => this.toggleDropdown(), "aria-haspopup": "listbox", "aria-expanded": this.isOpen ? 'true' : 'false', "aria-label": this.name, onKeyDown: e => this.handleSelectButtonKeyDown(e) }, h("slot", { key: '021e59628b8da260b308819614297c177a666334', name: "select-icon", "aria-hidden": "true" }), h("span", { key: '4a8c921538a23abe4c7efbeb176fb806587092c9', class: "selected-option", part: "msd_dropdown-button-selected-label" }, this.getSelectedOptions().length > 1
            ? `Multiple ${this.name}s selected`
            : this.availableOptions.filter(o => o.selected)[0]?.label ?? this.availableOptions[0]?.label), h("span", { key: 'fc771549d102fb27e33d128504d09fa6151154e9', class: this.isOpen ? 'open' : 'closed', "aria-hidden": "true" }, h("chevron-icon", { key: 'bd8282695d11bcfb617da05a75d0a2335650f705', width: "16", height: "16" }))), this.isOpen ? null : (h("div", { class: 'selected-options' }, this.getSelectedOptions().map(option => (h("span", { class: "selected-option" }, option.label, h("button", { key: option.label, onClick: () => this.selectOption(option) }, h("close-icon", null))))))), this.isOpen ? (h("div", { class: "dropdown-content", part: "msd_dropdown-content" }, h("ul", { tabindex: "-1", role: "listbox", "aria-label": this.name, "aria-multiselectable": true, "aria-activedescendant": this.ariaActivedescendant, onKeyDown: e => this.handleListboxKeydown(e) }, this.availableOptions.map(option => this.renderOption(option))))) : null)));
    }
    get el() { return this; }
    static get style() { return multiSelectDropdownCss; }
}, [1, "multi-select-dropdown", {
        "name": [1],
        "label": [1],
        "options": [16],
        "availableOptions": [32],
        "isOpen": [32],
        "ariaActivedescendant": [32],
        "shouldFocusFirstOption": [32]
    }, [[6, "click", "handleOutsideClick"]]]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["multi-select-dropdown", "chevron-icon", "close-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "multi-select-dropdown":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MultiSelectDropdown);
            }
            break;
        case "chevron-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "close-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { MultiSelectDropdown as M, defineCustomElement as d };
