// @nylas/web-elements/dist/components/nylas-summarize-message-button2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { a as debug, E as htmlToPlainText } from './utils.js';
import { R as RegisterComponent } from './register-component.js';
import { d as defineCustomElement$3 } from './loading.js';
import { d as defineCustomElement$2 } from './play.js';
import { d as defineCustomElement$1 } from './stop.js';

const nylasSummarizeMessageButtonCss = ":host{display:block}sp-button{padding:0;border:0;border-radius:var(--nylas-border-radius);min-block-size:50%;min-inline-size:2rem;background-color:var(--nylas-color-primary-300);fill:var(--nylas-color-primary-800)}sp-button:hover{background-color:var(--nylas-color-primary-400)}.summarize-button.stopped{animation:none}.summarize-button.playing{animation:pulse 2s infinite}.summarize-button.loading{display:flex;align-content:center;height:1.5rem;width:2rem;vertical-align:middle;animation:pulse 2s infinite;flex-wrap:wrap}.summarize-button.loading loading-icon{--dot-color:var(--nylas-color-primary-600)}@keyframes pulse{0%{transform:scale(1);box-shadow:0 0 0 0 var(--nylas-color-primary-300)}70%{transform:scale(1.05);box-shadow:0 0 0 10px rgba(255, 69, 0, 0)}100%{transform:scale(1);box-shadow:0 0 0 0 rgba(255, 69, 0, 0)}}@keyframes wave{0%{transform:translateX(0)}50%{transform:translateX(100%)}100%{transform:translateX(0)}}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasSummarizeMessageButton = proxyCustomElement(class NylasSummarizeMessageButton extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.summarize = createEvent(this, "summarize", 7);
        this.abortController = new AbortController();
        this.onClick = async (event) => {
            event.preventDefault();
            event.stopPropagation();
            if (this.state === 'stopped') {
                if (this.message) {
                    this.abortController = new AbortController();
                    const event = this.summarize.emit({
                        abortController: this.abortController,
                        playAudio: this.playAudio,
                        message: this.message,
                        updateState: this.updateState,
                    });
                    if (!event.defaultPrevented) {
                        debug(`[nylas-summarize-message-button] Default event not prevented, summarizing message`);
                        if (this.audioRef) {
                            await this.audioRef
                                .play()
                                .then(() => this.audioRef?.pause())
                                .catch(error => console.error('Error playing audio:', error));
                            await this.audioRef.play().catch(error => console.error('Error playing audio:', error));
                        }
                    }
                }
            }
            else if (this.state === 'playing') {
                this.abortController?.abort();
                this.audioRef?.pause();
            }
            else {
                this.state = 'stopped';
            }
        };
        this.message = undefined;
        this.state = 'stopped';
        this.playAudio = this.playAudio.bind(this);
        this.updateState = this.updateState.bind(this);
    }
    connectedCallback() { }
    disconnectedCallback() { }
    async componentWillLoad() {
        debug(`[nylas-summarize-message-button] Component will load`);
    }
    async componentDidLoad() {
        debug(`[nylas-summarize-message-button] Component did load`);
    }
    async updateState(state) {
        this.state = state;
    }
    async playAudio(ttsResponseBody) {
        if (!this.audioRef) {
            this.state = 'stopped';
            return;
        }
        if (!ttsResponseBody) {
            this.state = 'stopped';
            return;
        }
        const audioElement = this.audioRef;
        if (!('MediaSource' in window)) {
            const rb = await ttsResponseBody;
            if (rb == null) {
                throw new Error('ReadableStream not yet supported in this browser.');
            }
            const reader = rb.getReader();
            const chunks = [];
            await this.readChunk(reader, chunks);
            const blob = new Blob(chunks, { type: 'audio/mpeg' });
            const blobUrl = URL.createObjectURL(blob);
            audioElement.src = blobUrl;
            return;
        }
        const mediaSource = new MediaSource();
        let sourceBuffer;
        let sourceDone = false;
        audioElement.src = URL.createObjectURL(mediaSource);
        mediaSource.addEventListener('sourceopen', onMediaSourceOpen);
        async function onMediaSourceOpen() {
            sourceBuffer = mediaSource.addSourceBuffer('audio/mpeg');
            sourceBuffer.addEventListener('updateend', onUpdateEnd);
            const rb = await ttsResponseBody;
            if (rb == null) {
                throw new Error('ReadableStream not yet supported in this browser.');
            }
            const reader = rb.getReader();
            const pump = () => {
                reader
                    .read()
                    .then(({ done, value }) => {
                    if (done) {
                        sourceDone = true;
                        return;
                    }
                    if (!sourceBuffer.updating) {
                        sourceBuffer.appendBuffer(value);
                    }
                    pump();
                })
                    .catch(error => {
                    console.error('Error while fetching and appending chunks:', error);
                });
            };
            pump();
            if (audioElement.paused) {
                audioElement.play().catch(error => console.error('Error playing audio:', error));
            }
        }
        function onUpdateEnd() {
            if (sourceDone) {
                mediaSource.endOfStream();
                mediaSource.removeEventListener('sourceopen', onMediaSourceOpen);
                mediaSource.removeEventListener('sourceclose', () => {
                    debug(`[nylas-summarize-message-button] Source closed`);
                });
                sourceBuffer.removeEventListener('updateend', onUpdateEnd);
                return;
            }
        }
    }
    async readChunk(reader, chunks) {
        return reader.read().then(({ done, value }) => {
            if (done) {
                return chunks;
            }
            chunks.push(value);
            return this.readChunk(reader, chunks);
        });
    }
    render() {
        return (h(Host, { key: '09d693158ce9480c000957afe31566d47f6fb5f2' }, h("sp-theme", { key: '21f246783a3d5786c1cee8f438e89947ee5f574d', scale: "medium", color: "dark" }, h("sp-button", { key: '7ef9c60895da162d3b521ddc7c070ff463e51094', label: "Summarize email", "icon-only": true, onClickCapture: this.onClick, class: {
                'summarize-button': true,
                'stopped': this.state === 'stopped',
                'loading': this.state === 'loading',
                'playing': this.state === 'playing',
            } }, this.state === 'stopped' && h("play-icon", null), this.state === 'loading' && h("loading-icon", null), this.state === 'playing' && h("stop-icon", null)), h("audio", { key: '7481ceeb62b6eb21c627d0e04be0e0e7a9e986ec', ref: r => (this.audioRef = r), autoPlay: true, onPlaying: () => (this.state = 'playing'), onPause: () => (this.state = 'stopped'), onEnded: () => {
                this.state = 'stopped';
                debug(`[nylas-summarize-message-button] Audio ended`);
            } }))));
    }
    static get style() { return nylasSummarizeMessageButtonCss; }
}, [1, "nylas-summarize-message-button", {
        "message": [16],
        "state": [32],
        "updateState": [64],
        "playAudio": [64]
    }]);
__decorate([
    RegisterComponent({
        name: 'nylas-summarize-message-button',
        eventToProps: {
            summarize: async (event, nylasConnector) => {
                const { playAudio, updateState, abortController, message } = event.detail;
                const text = htmlToPlainText(message.body || '');
                await updateState('loading');
                try {
                    const summary = await nylasConnector.messages.summarizeText(text);
                    const ttsResponseBody = nylasConnector.messages.textToSpeech(summary, abortController);
                    await playAudio(ttsResponseBody);
                }
                catch (error) {
                    await updateState('stopped');
                    console.error('Error summarizing message:', error);
                }
            },
        },
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasSummarizeMessageButton.prototype, "render", null);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["nylas-summarize-message-button", "loading-icon", "play-icon", "stop-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "nylas-summarize-message-button":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, NylasSummarizeMessageButton);
            }
            break;
        case "loading-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "play-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "stop-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { NylasSummarizeMessageButton as N, defineCustomElement as d };
