// @nylas/web-elements/dist/components/sent.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-sent-icon-h{display:flex}";

const SentIcon = proxyCustomElement(class SentIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("svg", { key: '94433264804ae9ea142d21a8e017c3551bcbfde1', width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "auto", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '69a7c5647e8930885a8c63f0c44e370524786087', d: "M21.5 12L3.5 3.5L3.5 20.5L21.5 12Z", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: '4b7bcce0e696c35b4334572d8de66852d64783d3', d: "M3.5 3.5L12 10.5", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: '8f4cb81790c663a5db23f33b7dbd8fd17f264a4e', d: "M3.5 20.5L12 13.5", stroke: "auto", "stroke-width": "1.5" })));
    }
    static get style() { return iconCss; }
}, [2, "sent-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["sent-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "sent-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, SentIcon);
            }
            break;
    } });
}

export { SentIcon as S, defineCustomElement as d };
