// @nylas/web-elements/dist/components/spam.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-spam-icon-h{display:flex}";

const SpamIcon = proxyCustomElement(class SpamIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("svg", { key: '05802e7eb5bcbb8e1e36e40a86dd465993a4646a', width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "auto", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '165cb535da9d8c8b44f064218b79045c318e9726', d: "M20 3H4C2.89543 3 2 3.89543 2 5V19C2 20.1046 2.89543 21 4 21H20C21.1046 21 22 20.1046 22 19V5C22 3.89543 21.1046 3 20 3Z", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: 'daddc1e02d766c833becc9af5592904e6a094594', d: "M2 5L12 12.5L22 5", stroke: "auto", "stroke-width": "1.5" })));
    }
    static get style() { return iconCss; }
}, [2, "spam-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["spam-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "spam-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, SpamIcon);
            }
            break;
    } });
}

export { SpamIcon as S, defineCustomElement as d };
