// @nylas/web-elements/dist/components/stop.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-stop-icon-h{display:flex}";

const StopIcon = proxyCustomElement(class StopIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("svg", { key: 'f6149723fe23e788bd2a2b7fb075fb929222cb31', width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "auto", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '4f6c453999391dae9f8271f341c9ffac4458fd79', d: "M4 4H20V20H4V4Z", stroke: "auto", "stroke-width": "1.5", "stroke-linecap": "round", "stroke-linejoin": "round" })));
    }
    static get style() { return iconCss; }
}, [2, "stop-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["stop-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "stop-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, StopIcon);
            }
            break;
    } });
}

export { StopIcon as S, defineCustomElement as d };
