// @nylas/web-elements/dist/components/trash.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const iconCss = ".sc-trash-icon-h{display:flex}";

const TrashIcon = proxyCustomElement(class TrashIcon extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.width = '15';
        this.height = '15';
    }
    render() {
        return (h("svg", { key: '0fa95af4a6b84258131efa3e098791621c7e242a', width: this.width, height: this.height, viewBox: "0 0 24 24", fill: "auto", xmlns: "http://www.w3.org/2000/svg" }, h("path", { key: '49c0591ec45279759c2a3f38e9c55851692a2283', d: "M3 6H5H21", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: '389e6eed81f978951fb2bdb0b594bb6187a55fc3', d: "M8 6V4C8 3.46957 8.21071 2.96086 8.58579 2.58579C8.96086 2.21071 9.46957 2 10 2H14C14.5304 2 15.0391 2.21071 15.4142 2.58579C15.7893 2.96086 16 3.46957 16 4V6M19 6V20C19 20.5304 18.7893 21.0391 18.4142 21.4142C18.0391 21.7893 17.5304 22 17 22H7C6.46957 22 5.96086 21.7893 5.58579 21.4142C5.21071 21.0391 5 20.5304 5 20V6H19Z", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: 'a961e902065b0a650b558afbe31b00f5d19945aa', d: "M10 11V17", stroke: "auto", "stroke-width": "1.5" }), h("path", { key: '491de36ad9620009c6646d795e40276da46401ab', d: "M14 11V17", stroke: "auto", "stroke-width": "1.5" })));
    }
    static get style() { return iconCss; }
}, [2, "trash-icon", {
        "width": [1],
        "height": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["trash-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "trash-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, TrashIcon);
            }
            break;
    } });
}

export { TrashIcon as T, defineCustomElement as d };
