// @nylas/web-elements/dist/collection/common/constants.js
import moment from "moment-timezone";
export const LANGUAGE_MAP = {
    'en-US': 'English',
};
export const TIMEZONE_MAP = (() => {
    const timeZones = moment.tz.names();
    return timeZones.reduce((acc, timeZone) => {
        acc[timeZone] = timeZone.replace(/\//g, ' / ');
        acc[timeZone] = acc[timeZone].replace(/_/g, ' ');
        return acc;
    }, {});
})();
export function getTimezoneOffset(timezone) {
    return moment.tz(timezone).format('Z');
}
export const EVENT_TITLE_TOKENS = [
    {
        token: '$invitee',
        value: '${invitee}',
        description: 'The name of the primary invitee.',
    },
    {
        token: '$invitee_email',
        value: '${invitee_email}',
        description: 'The email of the primary invitee.',
    },
    {
        token: '$duration',
        value: '${duration}',
        description: 'The length in time of the event.',
    },
];
export const DEFAULT_OPEN_HOURS = [
    {
        days: [1, 2, 3, 4, 5],
        ex_dates: [],
        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
        start: '09:00',
        end: '17:00',
    },
];
