// @nylas/web-elements/dist/collection/common/nylas-api-request.js
import { dataResponse } from "../utils/utils";
export class NylasApiRequest {
}
export class NylasIdentityRequestWrapper {
    constructor(session) {
        this.defaultAuthArgs = {};
        this.session = session;
    }
    setActiveGrantId(grantId) {
        this.activeGrantId = grantId;
    }
    async request({ method, path, body, headers }) {
        try {
            const response = await this.session.fetch(`grants/me/${path}`, {
                method,
                headers: headers ? new Headers(headers) : undefined,
                body,
                ...(this.session.isMultiAccount() && this.activeGrantId ? { grant_id: this.activeGrantId } : {}),
            });
            if (typeof response === 'undefined') {
                throw new Error('Response is undefined');
            }
            return dataResponse(response);
        }
        catch (error) {
            return dataResponse(null, {
                message: String(error),
            });
        }
    }
    async currentUser() {
        const grantId = this.session.isMultiAccount() ? this.activeGrantId : undefined;
        const idToken = await this.session.getProfile(grantId);
        if (!idToken) {
            return null;
        }
        return {
            id: idToken.sub,
            email: idToken.email,
            name: idToken.name,
        };
    }
    async authenticationUrl(authArgs) {
        let args = this.defaultAuthArgs || {};
        if (authArgs) {
            args = { ...args, ...authArgs };
        }
        return this.session.auth(args);
    }
    setDefaultAuthArgs(authArgs) {
        this.defaultAuthArgs = authArgs;
    }
}
