// @nylas/web-elements/dist/collection/components/scheduler/nylas-notification/nylas-notification.js
import { Fragment, Host, h } from "@stencil/core";
import { debug } from "../../../utils/utils";
export class NylasNotification {
    constructor() {
        this.ttl = 5000;
        this.allowedCategories = ['component', 'api', 'validation', 'invalid_session'];
        this.notifications = [];
    }
    onError(event) {
        debug(`[nylas-notification] Error: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onWarning(event) {
        debug(`[nylas-notification] Warning: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onInfo(event) {
        debug(`[nylas-notification] Info: ${event.detail.title}`);
        this.handleNotification(event);
    }
    onSuccess(event) {
        debug(`[nylas-notification] Success: ${event.detail.title}`);
        this.handleNotification(event);
    }
    handleNotification(event) {
        if (event.defaultPrevented) {
            debug(`[nylas-notification] Event default prevented`);
            return;
        }
        if (this.ttl === 0) {
            debug(`[nylas-notification] TTL is 0`);
            return;
        }
        if (this.allowedCategories.length > 0 && !this.allowedCategories.includes(event.detail.category)) {
            debug(`[nylas-notification] Category not allowed`);
            return;
        }
        this.notifications = [...this.notifications, { ...event.detail }];
        if (event.detail.ttl !== 'none') {
            const ttl = event.detail.ttl ? event.detail.ttl : this.ttl;
            window.setTimeout(() => {
                this.notifications = this.notifications.filter(notif => notif.id !== event.detail.id);
            }, ttl);
        }
    }
    dismissNotification(id) {
        this.notifications = this.notifications.filter(notif => notif.id !== id);
    }
    render() {
        return (h(Host, { key: '16bd673245819e3ed76ba5689c3292be510e3cd7' }, h(Fragment, { key: 'd88e8ea863d00a8c982c7e26007c975526c7457f' }, this.notifications.map(notif => (h("div", { class: `notif ${notif.type}`, key: notif.id }, h("div", { class: "content" }, h("span", { class: `icon ${notif.type}` }, (notif.type === 'error' || notif.type === 'warning') && h("warning-icon", null)), h("span", { class: "message" }, h("span", { class: "title" }, notif.title), notif.description && h("span", { class: "description" }, notif.description))), h("button", { class: "dismiss-btn", onClick: () => this.dismissNotification(notif.id) }, h("close-icon", null))))), h("slot", { key: 'c40cb61ce0a1dbd3416824b7c26e08e3aea01eae' }))));
    }
    static get is() { return "nylas-notification"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["nylas-notification.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["nylas-notification.css"]
        };
    }
    static get properties() {
        return {
            "ttl": {
                "type": "any",
                "mutable": false,
                "complexType": {
                    "original": "number | 'none'",
                    "resolved": "\"none\" | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The time-to-live for notifications in milliseconds. Set to 'none' to disable auto-dismissal."
                },
                "attribute": "ttl",
                "reflect": false,
                "defaultValue": "5000"
            },
            "allowedCategories": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "defaultValue": "['component', 'api', 'validation', 'invalid_session']"
            }
        };
    }
    static get states() {
        return {
            "notifications": {}
        };
    }
    static get listeners() {
        return [{
                "name": "nylasSchedulerError",
                "method": "onError",
                "target": "body",
                "capture": false,
                "passive": false
            }, {
                "name": "nylasSchedulerWarning",
                "method": "onWarning",
                "target": "body",
                "capture": false,
                "passive": false
            }, {
                "name": "nylasSchedulerInfo",
                "method": "onInfo",
                "target": "body",
                "capture": false,
                "passive": false
            }, {
                "name": "nylasSchedulerSuccess",
                "method": "onSuccess",
                "target": "body",
                "capture": false,
                "passive": false
            }];
    }
}
