// @nylas/web-elements/dist/esm/nylas-threads-search.entry.js
import { r as registerInstance, i as createEvent, h, a as Host } from './index-8362ce5c.js';
import { R as RegisterComponent } from './register-component-0462b27d.js';
import { a as debug } from './utils-157014b4.js';
import './_commonjsHelpers-7c16df4a.js';

const nylasThreadsSearchCss = ":host{display:block}sp-search{width:100%;border:0}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasThreadsSearch = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.search = createEvent(this, "search", 7);
        this.onSearch = async (event) => {
            event.preventDefault();
            event.stopPropagation();
            const target = event.target;
            const value = target.value;
            if (value === this.internalQuery) {
                return;
            }
            this.internalQuery = value;
            this.search.emit(this.internalQuery);
        };
        this.query = '';
        this.internalQuery = '';
    }
    connectedCallback() { }
    disconnectedCallback() { }
    async componentWillLoad() {
        debug(`[nylas-threads-search] Component will load`);
    }
    async componentDidLoad() {
        debug(`[nylas-threads-search] Component did load`);
    }
    onQueryChange() {
        this.internalQuery = this.query;
    }
    render() {
        return (h(Host, { key: '9e91a330bea4dc84fa96070591e95926f52ee0bb' }, h("sp-theme", { key: '6d1df81ced09a5b721d8f161e40450633ae2cd98', scale: "medium", color: "dark" }, h("sp-search", { key: 'cb81bba4b4f1d279f1fd48bda9a2fb003cb65fcd', quiet: true, onSubmit: this.onSearch, onBlur: this.onSearch, value: this.internalQuery }))));
    }
    static get watchers() { return {
        "query": ["onQueryChange"]
    }; }
};
__decorate([
    RegisterComponent({
        name: 'nylas-pagination',
        stateToProps: new Map([['mailbox.searchQuery', 'query']]),
        eventToProps: {
            search: async (_event, nylasConnector) => {
                await nylasConnector.threads.getThreads({
                    search_query_native: _event.detail || undefined,
                    page_token: undefined,
                });
            },
        },
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasThreadsSearch.prototype, "render", null);
NylasThreadsSearch.style = nylasThreadsSearchCss;

export { NylasThreadsSearch as nylas_threads_search };
