// @nylas/web-elements/dist/esm/nylas-event-limits.entry.js
import { r as registerInstance, h, a as Host } from './index-8362ce5c.js';
import { R as RegisterComponent } from './register-component-0462b27d.js';
import { a as debug } from './utils-157014b4.js';
import './_commonjsHelpers-7c16df4a.js';

const nylasEventLimitsCss = ":host{display:block;margin:1rem;--nylas-primary:#2563eb;--nylas-error:#cc4841;--nylas-error-pressed:#992222;--nylas-wraning:#f06c00;--nylas-success:#16a392;--nylas-info:#2b8fc2;--nylas-base-0:#ffffff;--nylas-base-25:#fcfcfd;--nylas-base-50:#f8f9fc;--nylas-base-100:#eaecf5;--nylas-base-200:#d5d9eb;--nylas-base-300:#b3b8d8;--nylas-base-400:#717bbc;--nylas-base-500:#4e5ba6;--nylas-base-600:#3e4784;--nylas-base-700:#263f72;--nylas-base-800:#293056;--nylas-base-900:#101323;--nylas-base-950:#0e101b;--nylas-border-radius:0.25rem;--nylas-border-radius-2x:0.5rem;--nylas-border-radius-3x:0.75rem;--nylas-font-family:\"Inter\", sans-serif}.nylas-event-limits__title{color:var(--nylas-base-900);font-size:16px;font-style:normal;font-weight:600;line-height:20px;margin:0;text-align:left}.nylas-event-limits__subtitle{color:var(--nylas-base-600);font-size:14px;font-style:normal;font-weight:400;line-height:20px;margin:0.25rem 0px 0px;text-align:left}.nylas-event-limits__content{padding:1rem;display:flex;flex-direction:column;gap:1rem}.nylas-event-limits__block{display:flex;justify-content:space-between;gap:0.5rem}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasEventLimits = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        if (hostRef.$hostElement$["s-ei"]) {
            this.internals = hostRef.$hostElement$["s-ei"];
        }
        else {
            this.internals = hostRef.$hostElement$.attachInternals();
            hostRef.$hostElement$["s-ei"] = this.internals;
        }
    }
    connectedCallback() {
        debug('nylas-event-limits', 'connectedCallback');
    }
    componentWillLoad() {
        debug('nylas-event-limits', 'componentWillLoad');
    }
    componentDidLoad() {
        debug('nylas-event-limits', 'componentDidLoad');
    }
    disconnectedCallback() {
        debug('nylas-event-limits', 'disconnectedCallback');
    }
    render() {
        return (h(Host, { key: '56a1396f87f4d81f1a3912ec810e23a9dbed965b' }, h("nylas-form-card", { key: '2007e2fe058e39bb8c276d2995b4b8075d9ed82f' }, h("h1", { key: 'e57bc40b4b369d64cdb2b6c97009dea3f141716f', slot: "header-title", class: "nylas-event-limits__title" }, "Set event limits"), h("h3", { key: 'beb61ca97cfd3dca960e78e5f64abca4ace7777b', slot: "header-subtitle", class: "nylas-event-limits__subtitle" }, "Customize event limits for the event."), h("div", { key: '3912501d3e927fcbc6e8663f9149c66ce54fbd0e', slot: "content", class: "nylas-event-limits__content" }, h("slot", { key: 'a0902c2edc5ef540547452d5f1448ba26547af17', name: "inputs" })))));
    }
    static get formAssociated() { return true; }
};
__decorate([
    RegisterComponent({
        name: 'nylas-event-limits',
        eventToProps: {},
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasEventLimits.prototype, "render", null);
NylasEventLimits.style = nylasEventLimitsCss;

export { NylasEventLimits as nylas_event_limits };
