// @nylas/web-elements/dist/esm/nylas-mailbox-toolbar-button.entry.js
import { r as registerInstance, h, a as Host, e as getElement } from './index-8362ce5c.js';
import { R as RegisterComponent } from './register-component-0462b27d.js';
import { a as debug } from './utils-157014b4.js';
import './_commonjsHelpers-7c16df4a.js';

const nylasMailboxToolbarButtonCss = ":host{display:block}sp-button{border-radius:var(--nylas-border-radius);background-color:var(--nylas-color-primary-400);color:var(--nylas-color-primary-900);fill:var(--nylas-color-primary-900)}sp-button:hover{background-color:var(--nylas-color-primary-500)}sp-button:disabled{background-color:var(--nylas-color-primary-200);color:var(--nylas-color-primary-600);cursor:not-allowed}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (undefined && undefined.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
        return Reflect.metadata(k, v);
};
const NylasMailboxToolbarButton = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.onClick = async () => {
            const method = this.method;
            debug(`[nylas-mailbox-toolbar-button] onClick`, { method, mailboxConnector: this.mailboxConnector });
            if (this.mailboxConnector && method) {
                const methodToCall = this.callMethod(this.mailboxConnector, method);
                if (methodToCall) {
                    await methodToCall();
                }
            }
        };
        this.method = undefined;
        this.dataState = undefined;
        this.iconOnly = false;
        this.mailboxConnector = undefined;
    }
    connectedCallback() { }
    disconnectedCallback() { }
    async componentWillLoad() {
        debug(`[nylas-if-state] Component will load`);
    }
    async componentDidLoad() {
        debug(`[nylas-if-state] Component did load`);
    }
    callMethod(obj, accessString) {
        const parts = accessString.split('.');
        let current = obj;
        for (const part of parts.slice(0, -1)) {
            current = current[part];
            if (!current) {
                return undefined;
            }
        }
        const methodName = parts[parts.length - 1];
        const method = current[methodName];
        if (typeof method === 'function') {
            return method.bind(current);
        }
        return undefined;
    }
    renderByState(state) {
        return state ? h("slot", { name: "loading" }, "Loading...") : h("slot", null, "Click");
    }
    hasSlot(name) {
        return !!this.host.querySelector(`[slot="${name}"]`);
    }
    render() {
        let state = false;
        const store = this.mailboxConnector?.nylasMailboxStore;
        if (typeof store !== 'undefined') {
            const storeKey = String(this.dataState);
            state = this.dataState ? store.get(storeKey) === 'loading' : false;
        }
        const key = state ? `state-loading` : 'state-ready';
        return (h(Host, { key: '2454ef2e4270d1642cabe57dacf53908fcdb34e6' }, h("sp-theme", { key: '97d71094fabc3466b87142dfdd1d0bb22659fd32', scale: "medium", color: "dark" }, h("sp-button", { onClick: this.onClick, disabled: state, key: key, "icon-only": this.iconOnly, label: this.renderByState(state) }, this.hasSlot('icon') && (h("div", { slot: "icon", class: 'slot' }, h("slot", { name: "icon" }))), !this.iconOnly && this.renderByState(state)))));
    }
    get host() { return getElement(this); }
};
__decorate([
    RegisterComponent({
        name: 'nylas-mailbox-toolbar-button',
        connectorToProp: 'mailboxConnector',
        fireRegisterEvent: true,
    }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NylasMailboxToolbarButton.prototype, "render", null);
NylasMailboxToolbarButton.style = nylasMailboxToolbarButtonCss;

export { NylasMailboxToolbarButton as nylas_mailbox_toolbar_button };
